// naive-ui/lib/_internal/menu-mask/src/MenuMask.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _mixins_1 = require("../../../_mixins");
const index_cssr_1 = require("./styles/index.cssr");
exports.default = vue_1.defineComponent({
    name: 'BaseMenuMask',
    props: {
        clsPrefix: {
            type: String,
            required: true
        }
    },
    setup(props) {
        _mixins_1.useStyle('BaseMenuMask', index_cssr_1.default, vue_1.toRef(props, 'clsPrefix'));
        const messageRef = vue_1.ref(null);
        let timerId = null;
        const uncontrolledShowRef = vue_1.ref(false);
        vue_1.onBeforeUnmount(() => {
            if (timerId !== null) {
                window.clearTimeout(timerId);
            }
        });
        const exposedRef = {
            showOnce(message, duration = 1500) {
                if (timerId)
                    window.clearTimeout(timerId);
                uncontrolledShowRef.value = true;
                messageRef.value = message;
                timerId = window.setTimeout(() => {
                    uncontrolledShowRef.value = false;
                    messageRef.value = null;
                }, duration);
            }
        };
        return Object.assign({ message: messageRef, show: uncontrolledShowRef }, exposedRef);
    },
    render() {
        return (vue_1.h(vue_1.Transition, { name: "fade-in-transition" }, {
            default: () => this.show ? (vue_1.h("div", { class: `${this.clsPrefix}-base-menu-mask` }, this.message)) : null
        }));
    }
});
