// naive-ui/lib/date-picker/src/panel/use-dual-calendar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useDualCalendar = void 0;
const vue_1 = require("vue");
const date_fns_1 = require("date-fns");
const utils_1 = require("../utils");
const use_panel_common_1 = require("./use-panel-common");
const interface_1 = require("../interface");
const useDualCalendarProps = Object.assign(Object.assign({}, use_panel_common_1.usePanelCommon.props), { actions: {
        type: Array,
        default: () => ['clear', 'confirm']
    } });
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
function useDualCalendar(props, type = 'datetime') {
    const { isDateDisabledRef, isStartHourDisabledRef, isEndHourDisabledRef, isStartMinuteDisabledRef, isEndMinuteDisabledRef, isStartSecondDisabledRef, isEndSecondDisabledRef, isStartDateInvalidRef, isEndDateInvalidRef, isStartTimeInvalidRef, isEndTimeInvalidRef, isStartValueInvalidRef, isEndValueInvalidRef, isRangeInvalidRef, localeRef, rangesRef, datePickerSlots
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
     } = vue_1.inject(interface_1.datePickerInjectionKey);
    const validation = {
        isDateDisabled: isDateDisabledRef,
        isStartHourDisabled: isStartHourDisabledRef,
        isEndHourDisabled: isEndHourDisabledRef,
        isStartMinuteDisabled: isStartMinuteDisabledRef,
        isEndMinuteDisabled: isEndMinuteDisabledRef,
        isStartSecondDisabled: isStartSecondDisabledRef,
        isEndSecondDisabled: isEndSecondDisabledRef,
        isStartDateInvalid: isStartDateInvalidRef,
        isEndDateInvalid: isEndDateInvalidRef,
        isStartTimeInvalid: isStartTimeInvalidRef,
        isEndTimeInvalid: isEndTimeInvalidRef,
        isStartValueInvalid: isStartValueInvalidRef,
        isEndValueInvalid: isEndValueInvalidRef,
        isRangeInvalid: isRangeInvalidRef
    };
    const panelCommon = use_panel_common_1.usePanelCommon(props);
    const startDatesElRef = vue_1.ref(null);
    const endDatesElRef = vue_1.ref(null);
    const startCalendarDateTimeRef = vue_1.ref(Date.now());
    const endCalendarDateTimeRef = vue_1.ref(date_fns_1.getTime(date_fns_1.addMonths(Date.now(), 1)));
    const nowRef = vue_1.ref(Date.now());
    const isSelectingRef = vue_1.ref(false);
    const memorizedStartDateTimeRef = vue_1.ref(0);
    const { value, dateFormat } = props;
    const startDateInput = vue_1.ref(Array.isArray(value)
        ? date_fns_1.format(value[0], dateFormat, panelCommon.dateFnsOptions.value)
        : '');
    const endDateInputRef = vue_1.ref(Array.isArray(value)
        ? date_fns_1.format(value[1], dateFormat, panelCommon.dateFnsOptions.value)
        : '');
    if (Array.isArray(value)) {
        syncCalendarTimeWithValue(value);
    }
    // derived computed
    const selectingPhaseRef = vue_1.computed(() => {
        if (isSelectingRef.value)
            return 'end';
        else
            return 'start';
    });
    const startDateArrayRef = vue_1.computed(() => {
        return utils_1.dateArray(startCalendarDateTimeRef.value, props.value, nowRef.value, localeRef.value.firstDayOfWeek);
    });
    const endDateArrayRef = vue_1.computed(() => {
        return utils_1.dateArray(endCalendarDateTimeRef.value, props.value, nowRef.value, localeRef.value.firstDayOfWeek);
    });
    const weekdaysRef = vue_1.computed(() => {
        return startDateArrayRef.value.slice(0, 7).map((dateItem) => {
            const { ts } = dateItem;
            return date_fns_1.format(ts, localeRef.value.dayFormat, panelCommon.dateFnsOptions.value);
        });
    });
    const startCalendarMonthRef = vue_1.computed(() => {
        return date_fns_1.format(startCalendarDateTimeRef.value, localeRef.value.monthFormat, panelCommon.dateFnsOptions.value);
    });
    const endCalendarMonthRef = vue_1.computed(() => {
        return date_fns_1.format(endCalendarDateTimeRef.value, localeRef.value.monthFormat, panelCommon.dateFnsOptions.value);
    });
    const startCalendarYearRef = vue_1.computed(() => {
        return date_fns_1.format(startCalendarDateTimeRef.value, localeRef.value.yearFormat, panelCommon.dateFnsOptions.value);
    });
    const endCalendarYearRef = vue_1.computed(() => {
        return date_fns_1.format(endCalendarDateTimeRef.value, localeRef.value.yearFormat, panelCommon.dateFnsOptions.value);
    });
    const startTimeValueRef = vue_1.computed(() => {
        const { value } = props;
        if (Array.isArray(value))
            return value[0];
        return null;
    });
    const endTimeValueRef = vue_1.computed(() => {
        const { value } = props;
        if (Array.isArray(value))
            return value[1];
        return null;
    });
    vue_1.watch(vue_1.computed(() => props.value), (value) => {
        if (value !== null && Array.isArray(value)) {
            const [startMoment, endMoment] = value;
            const { dateFormat } = props;
            startDateInput.value = date_fns_1.format(startMoment, dateFormat, panelCommon.dateFnsOptions.value);
            endDateInputRef.value = date_fns_1.format(endMoment, dateFormat, panelCommon.dateFnsOptions.value);
            if (!isSelectingRef.value) {
                syncCalendarTimeWithValue(value);
            }
        }
        else {
            startDateInput.value = '';
            endDateInputRef.value = '';
        }
    });
    function handleCalendarChange(value, oldValue) {
        if (date_fns_1.getYear(value) !== date_fns_1.getYear(oldValue) ||
            date_fns_1.getMonth(value) !== date_fns_1.getMonth(oldValue)) {
            panelCommon.disableTransitionOneTick();
        }
    }
    vue_1.watch(startCalendarDateTimeRef, handleCalendarChange);
    vue_1.watch(endCalendarDateTimeRef, handleCalendarChange);
    // change calendar
    function adjustCalendarTimes(byStartCalendarTime) {
        const startTime = date_fns_1.startOfMonth(startCalendarDateTimeRef.value);
        const endTime = date_fns_1.startOfMonth(endCalendarDateTimeRef.value);
        if (startTime >= endTime) {
            if (byStartCalendarTime) {
                endCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(startTime, 1));
            }
            else {
                startCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(endTime, -1));
            }
        }
    }
    function startCalendarNextYear() {
        startCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(startCalendarDateTimeRef.value, 12));
        adjustCalendarTimes(true);
    }
    function startCalendarPrevYear() {
        startCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(startCalendarDateTimeRef.value, -12));
        adjustCalendarTimes(true);
    }
    function startCalendarNextMonth() {
        startCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(startCalendarDateTimeRef.value, 1));
        adjustCalendarTimes(true);
    }
    function startCalendarPrevMonth() {
        startCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(startCalendarDateTimeRef.value, -1));
        adjustCalendarTimes(true);
    }
    function endCalendarNextYear() {
        endCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(endCalendarDateTimeRef.value, 12));
        adjustCalendarTimes(false);
    }
    function endCalendarPrevYear() {
        endCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(endCalendarDateTimeRef.value, -12));
        adjustCalendarTimes(false);
    }
    function endCalendarNextMonth() {
        endCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(endCalendarDateTimeRef.value, 1));
        adjustCalendarTimes(false);
    }
    function endCalendarPrevMonth() {
        endCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.addMonths(endCalendarDateTimeRef.value, -1));
        adjustCalendarTimes(false);
    }
    // The function is used on date panel, not the date-picker value validation
    function mergedIsDateDisabled(ts) {
        const isDateDisabled = isDateDisabledRef.value;
        if (!isDateDisabled)
            return false;
        if (!Array.isArray(props.value))
            return isDateDisabled(ts, 'start', null);
        if (selectingPhaseRef.value === 'start') {
            // before you really start to select
            return isDateDisabled(ts, 'start', null);
        }
        else {
            const { value: memorizedStartDateTime } = memorizedStartDateTimeRef;
            // after you starting to select
            if (ts < memorizedStartDateTimeRef.value) {
                return isDateDisabled(ts, 'start', [
                    memorizedStartDateTime,
                    memorizedStartDateTime
                ]);
            }
            else {
                return isDateDisabled(ts, 'end', [
                    memorizedStartDateTime,
                    memorizedStartDateTime
                ]);
            }
        }
    }
    function resetSelectingStatus(e) {
        var _a, _b;
        if (((_a = startDatesElRef.value) === null || _a === void 0 ? void 0 : _a.contains(e.target)) ||
            ((_b = endDatesElRef.value) === null || _b === void 0 ? void 0 : _b.contains(e.target))) {
            // do nothing
        }
        else {
            isSelectingRef.value = false;
        }
    }
    function syncCalendarTimeWithValue(value) {
        if (value === null)
            return;
        const [startMoment, endMoment] = value;
        startCalendarDateTimeRef.value = startMoment;
        if (date_fns_1.startOfMonth(endMoment) <= date_fns_1.startOfMonth(startMoment)) {
            endCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.startOfMonth(date_fns_1.addMonths(startMoment, 1)));
        }
        else {
            endCalendarDateTimeRef.value = date_fns_1.getTime(date_fns_1.startOfMonth(endMoment));
        }
    }
    function handleDateClick(dateItem) {
        if (mergedIsDateDisabled(dateItem.ts)) {
            return;
        }
        if (!isSelectingRef.value) {
            isSelectingRef.value = true;
            memorizedStartDateTimeRef.value = dateItem.ts;
            changeStartEndTime(dateItem.ts);
        }
        else {
            isSelectingRef.value = false;
        }
    }
    function handleDateMouseEnter(dateItem) {
        if (isSelectingRef.value) {
            if (mergedIsDateDisabled(dateItem.ts))
                return;
            if (dateItem.ts >= memorizedStartDateTimeRef.value) {
                changeStartEndTime(memorizedStartDateTimeRef.value, dateItem.ts);
            }
            else {
                changeStartEndTime(dateItem.ts, memorizedStartDateTimeRef.value);
            }
        }
    }
    function handleConfirmClick() {
        if (isRangeInvalidRef.value) {
            return;
        }
        panelCommon.doConfirm();
        closeCalendar();
    }
    function closeCalendar() {
        isSelectingRef.value = false;
        if (props.active) {
            panelCommon.doClose();
        }
    }
    function changeStartDateTime(time) {
        if (typeof time !== 'number') {
            time = date_fns_1.getTime(time);
        }
        if (props.value === null) {
            panelCommon.doUpdateValue([time, time], false);
        }
        else if (Array.isArray(props.value)) {
            panelCommon.doUpdateValue([time, Math.max(props.value[1], time)], false);
        }
    }
    function changeEndDateTime(time) {
        if (typeof time !== 'number') {
            time = date_fns_1.getTime(time);
        }
        if (props.value === null) {
            panelCommon.doUpdateValue([time, time], false);
        }
        else if (Array.isArray(props.value)) {
            panelCommon.doUpdateValue([Math.min(props.value[0], time), time], false);
        }
    }
    let cachedValue = null;
    let cached = false;
    function cachePendingValue() {
        cachedValue = props.value;
        cached = true;
    }
    function clearPendingValue() {
        cached = false;
    }
    function restorePendingValue() {
        if (cached) {
            panelCommon.doUpdateValue(cachedValue, false);
            cached = false;
        }
    }
    function changeStartEndTime(startTime, endTime) {
        if (endTime === undefined)
            endTime = startTime;
        if (typeof startTime !== 'number') {
            startTime = date_fns_1.getTime(startTime);
        }
        if (typeof endTime !== 'number') {
            endTime = date_fns_1.getTime(endTime);
        }
        panelCommon.doUpdateValue([startTime, endTime], false);
    }
    function sanitizeValue(datetime) {
        if (type === 'datetimerange') {
            return date_fns_1.getTime(date_fns_1.startOfSecond(datetime));
        }
        else {
            // daterange
            return date_fns_1.getTime(date_fns_1.startOfDay(datetime));
        }
    }
    function handleStartDateInput(value) {
        const date = utils_1.strictParse(value, props.dateFormat, new Date(), panelCommon.dateFnsOptions.value);
        if (date_fns_1.isValid(date)) {
            if (!props.value) {
                const newValue = date_fns_1.set(new Date(), {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeStartDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
            else if (Array.isArray(props.value)) {
                const newValue = date_fns_1.set(props.value[0], {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeStartDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
        }
        else {
            startDateInput.value = value;
        }
    }
    function handleEndDateInput(value) {
        /** strict check when input */
        const date = utils_1.strictParse(value, props.dateFormat, new Date(), panelCommon.dateFnsOptions.value);
        if (date_fns_1.isValid(date)) {
            if (props.value === null) {
                const newValue = date_fns_1.set(new Date(), {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeEndDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
            else if (Array.isArray(props.value)) {
                const newValue = date_fns_1.set(props.value[1], {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeEndDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
        }
        else {
            endDateInputRef.value = value;
        }
    }
    function handleStartDateInputBlur() {
        const date = utils_1.strictParse(startDateInput.value, props.dateFormat, new Date(), panelCommon.dateFnsOptions.value);
        const { value } = props;
        if (date_fns_1.isValid(date)) {
            if (value === null) {
                const newValue = date_fns_1.set(new Date(), {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeStartDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
            else if (Array.isArray(value)) {
                const newValue = date_fns_1.set(value[0], {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeStartDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
        }
        else {
            refreshDisplayDateString();
        }
    }
    function handleEndDateInputBlur() {
        const date = utils_1.strictParse(endDateInputRef.value, props.dateFormat, new Date(), panelCommon.dateFnsOptions.value);
        const { value } = props;
        if (date_fns_1.isValid(date)) {
            if (value === null) {
                const newValue = date_fns_1.set(new Date(), {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeEndDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
            else if (Array.isArray(value)) {
                const newValue = date_fns_1.set(value[1], {
                    year: date_fns_1.getYear(date),
                    month: date_fns_1.getMonth(date),
                    date: date_fns_1.getDate(date)
                });
                changeEndDateTime(sanitizeValue(date_fns_1.getTime(newValue)));
            }
        }
        else {
            refreshDisplayDateString();
        }
    }
    function refreshDisplayDateString(times) {
        // If not selected, display nothing,
        // else update datetime related string
        const { value } = props;
        if (value === null || !Array.isArray(value)) {
            startDateInput.value = '';
            endDateInputRef.value = '';
            return;
        }
        const { dateFormat } = props;
        if (times === undefined) {
            times = value;
        }
        startDateInput.value = date_fns_1.format(times[0], dateFormat, panelCommon.dateFnsOptions.value);
        endDateInputRef.value = date_fns_1.format(times[1], dateFormat, panelCommon.dateFnsOptions.value);
    }
    function handleStartTimePickerChange(value) {
        changeStartDateTime(value);
    }
    function handleEndTimePickerChange(value) {
        changeEndDateTime(value);
    }
    return Object.assign(Object.assign(Object.assign({ startDatesElRef,
        endDatesElRef,
        resetSelectingStatus,
        handleDateClick,
        handleDateMouseEnter,
        handleConfirmClick,
        startCalendarPrevYear,
        startCalendarPrevMonth,
        startCalendarNextYear,
        startCalendarNextMonth,
        endCalendarPrevYear,
        endCalendarPrevMonth,
        endCalendarNextMonth,
        endCalendarNextYear,
        mergedIsDateDisabled,
        changeStartEndTime,
        cachePendingValue,
        clearPendingValue,
        restorePendingValue, ranges: rangesRef, startCalendarMonth: startCalendarMonthRef, startCalendarYear: startCalendarYearRef, endCalendarMonth: endCalendarMonthRef, endCalendarYear: endCalendarYearRef, weekdays: weekdaysRef, startDateArray: startDateArrayRef, endDateArray: endDateArrayRef }, panelCommon), validation), { 
        // datetimerangeonly
        startDateDisplayString: startDateInput, endDateInput: endDateInputRef, timePickerSize: panelCommon.timePickerSize, startTimeValue: startTimeValueRef, endTimeValue: endTimeValueRef, handleFocusDetectorFocus: panelCommon.handleFocusDetectorFocus, handleStartTimePickerChange,
        handleEndTimePickerChange,
        handleStartDateInput,
        handleStartDateInputBlur,
        handleEndDateInput,
        handleEndDateInputBlur,
        datePickerSlots });
}
exports.useDualCalendar = useDualCalendar;
useDualCalendar.props = useDualCalendarProps;
