// naive-ui/es/message/src/MessageEnvironment.js
import { h, defineComponent, ref, onMounted } from 'vue';
import { NFadeInExpandTransition } from '../../_internal';
import NMessage from './Message';
import { messageProps } from './message-props';
export default defineComponent({
    name: 'MessageEnvironment',
    props: Object.assign(Object.assign({}, messageProps), { duration: {
            type: Number,
            default: 3000
        }, onAfterLeave: Function, onLeave: Function, internalKey: {
            type: String,
            required: true
        }, 
        // private
        onInternalAfterLeave: Function, 
        // deprecated
        onHide: Function, onAfterHide: Function }),
    setup(props) {
        const timerIdRef = ref(null);
        const showRef = ref(true);
        onMounted(() => {
            const { duration } = props;
            if (duration) {
                timerIdRef.value = window.setTimeout(hide, duration);
            }
        });
        function hide() {
            const { value: timerId } = timerIdRef;
            const { onHide } = props;
            showRef.value = false;
            if (timerId) {
                window.clearTimeout(timerId);
            }
            // deprecated
            if (onHide)
                onHide();
        }
        function handleClose() {
            const { onClose } = props;
            if (onClose)
                onClose();
            hide();
        }
        function handleAfterLeave() {
            const { onAfterLeave, onInternalAfterLeave, onAfterHide, internalKey } = props;
            if (onAfterLeave)
                onAfterLeave();
            if (onInternalAfterLeave)
                onInternalAfterLeave(internalKey);
            // deprecated
            if (onAfterHide)
                onAfterHide();
        }
        // deprecated
        function deactivate() {
            hide();
        }
        return {
            show: showRef,
            hide,
            handleClose,
            handleAfterLeave,
            deactivate
        };
    },
    render() {
        return (h(NFadeInExpandTransition, { appear: true, onAfterLeave: this.handleAfterLeave, onLeave: this.onLeave }, {
            default: () => [
                this.show ? (h(NMessage, { content: this.content, type: this.type, icon: this.icon, closable: this.closable, onClose: this.handleClose })) : null
            ]
        }));
    }
});
