// naive-ui/lib/upload/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_height_expand_cssr_1 = require("../../../_styles/transitions/fade-in-height-expand.cssr");

const icon_switch_cssr_1 = require("../../../_styles/transitions/icon-switch.cssr");

exports.default = cssr_1.cB('upload', [cssr_1.cE('file-input', `
 display: block;
 width: 0;
 height: 0;
 opacity: 0;
 `), cssr_1.cE('trigger', `
 display: inline-block;
 `), cssr_1.cM('dragger-inside', [cssr_1.cE('trigger', `
 display: block;
 `)]), cssr_1.cB('upload-dragger', `
 cursor: pointer;
 box-sizing: border-box;
 width: 100%;
 text-align: center;
 border-radius: var(--border-radius);
 padding: 24px;
 transition:
 border-color .3s var(--bezier),
 background-color .3s var(--bezier);
 background-color: var(--dragger-color);
 border: var(--dragger-border);
 `, [cssr_1.c('&:hover', `
 border: var(--dragger-border-hover);
 `)]), cssr_1.cB('upload-file-list', `
 margin-top: 8px;
 line-height: var(--line-height);
 `, [cssr_1.cB('upload-file', `
 display: block;
 box-sizing: border-box;
 cursor: default;
 padding: 0px 12px 0 6px;
 transition: background-color .3s var(--bezier);
 border-radius: var(--border-radius);
 `, [fade_in_height_expand_cssr_1.default(), cssr_1.cB('progress', `
 box-sizing: border-box;
 padding-bottom: 6px;
 margin-bottom: 6px;
 `, [fade_in_height_expand_cssr_1.default({
  foldPadding: true
})]), cssr_1.c('&:hover', `
 background-color: var(--item-color-hover);
 `, [cssr_1.cB('upload-file-info', [cssr_1.cE('action', `
 opacity: 1;
 `)])]), cssr_1.cM('error-status', [cssr_1.c('&:hover', `
 background-color: var(--item-color-hover-error);
 `), cssr_1.cB('upload-file-info', [cssr_1.cE('name', `
 color: var(--item-text-color-error);
 `)])]), cssr_1.cM('with-url', `
 cursor: pointer;
 `, [cssr_1.cB('upload-file-info', [cssr_1.cE('name', `
 text-decoration: underline;
 color: var(--item-text-color-success);
 text-decoration-color: var(--item-text-color-success);
 `)])]), cssr_1.cB('upload-file-info', `
 position: relative;
 padding-top: 6px;
 padding-bottom: 6px;
 `, [cssr_1.cE('action', `
 padding-top: inherit;
 padding-bottom: inherit;
 position: absolute;
 right: 0;
 top: 0;
 bottom: 0;
 width: 80px;
 display: flex;
 align-items: center;
 transition: opacity .2s var(--bezier);
 justify-content: flex-end;
 opacity: 0;
 `, [cssr_1.cB('button', [cssr_1.c('&:not(:last-child)', {
  marginRight: '4px'
}), cssr_1.cB('base-icon', [cssr_1.c('svg', [icon_switch_cssr_1.default()])])])]), cssr_1.cE('name', `
 display: flex;
 align-items: center;
 text-overflow: ellipsis;
 overflow: hidden;
 text-decoration: underline;
 text-decoration-color: #0000;
 font-size: var(--font-size);
 transition:
 color .3s var(--bezier),
 text-decoration-color .3s var(--bezier);
 color: var(--item-text-color); 
 `, [cssr_1.cB('base-icon', `
 font-size: 18px;
 margin-right: 2px;
 vertical-align: middle;
 color: var(--item-icon-color);
 `)])])])]), cssr_1.cM('disabled', `
 opacity: var(--item-disabled-opacity);
 `, [cssr_1.cE('trigger', `
 cursor: not-allowed;
 `), cssr_1.cB('upload-file-list', `
 cursor: not-allowed;
 `), cssr_1.cB('upload-dragger', `
 cursor: not-allowed;
 `)]), cssr_1.cM('drag-over', [cssr_1.cB('upload-dragger', `
 border: var(--dragger-border-hover);
 `)])]);
