// naive-ui/lib/collapse/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_height_expand_cssr_1 = require("../../../_styles/transitions/fade-in-height-expand.cssr"); // vars:
// --font-size
// --bezier
// --text-color
// --divider-color
// --title-font-size
// --title-text-color
// --title-font-weight
// --arrow-color


exports.default = cssr_1.cB('collapse', {
  width: '100%'
}, [cssr_1.cB('collapse-item', `
 font-size: var(--font-size);
 color: var(--text-color);
 transition:
 color .3s var(--bezier),
 border-color .3s var(--bezier);
 margin: 16px 0 0 0;
 `, [cssr_1.cB('collapse-item', 'margin-left: 32px;'), cssr_1.c('&:first-child', {
  marginTop: 0
}), cssr_1.c('&:first-child >', [cssr_1.cE('header', {
  paddingTop: 0
})]), cssr_1.cM('left-arrow-placement', [cssr_1.cE('header', [cssr_1.cB('collapse-item-arrow', {
  marginRight: '4px'
})])]), cssr_1.cM('right-arrow-placement', [cssr_1.cE('header', [cssr_1.cB('collapse-item-arrow', {
  marginLeft: '4px'
})])]), cssr_1.cE('content-wrapper', [cssr_1.cE('content-inner', {
  paddingTop: '16px'
}), fade_in_height_expand_cssr_1.default({
  duration: '0.15s'
})]), cssr_1.cM('active', [cssr_1.cE('header', [cssr_1.cM('active', [cssr_1.cB('collapse-item-arrow', {
  transform: 'rotate(90deg)'
})])])]), cssr_1.c('&:not(:first-child)', {
  borderTop: '1px solid var(--divider-color)'
}), cssr_1.cE('header', `
 font-size: var(--title-font-size);
 display: flex;
 flex-wrap: nowrap;
 align-items: center;
 font-weight: var(--title-font-weight);
 transition: color .3s var(--bezier);
 position: relative;
 cursor: pointer;
 padding: 16px 0 0 0;
 color: var(--title-text-color);
 `, [cssr_1.cB('collapse-item-arrow', `
 display: flex;
 transition:
 transform .15s var(--bezier),
 color .3s var(--bezier);
 font-size: 18px;
 color: var(--arrow-color);
 `)])])]);
