// naive-ui/lib/notification/src/NotificationContainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const scrollbar_1 = require("../../scrollbar");
const NotificationProvider_1 = require("./NotificationProvider");
exports.default = vue_1.defineComponent({
    name: 'NotificationContainer',
    props: {
        scrollable: {
            type: Boolean,
            required: true
        }
    },
    setup() {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const { mergedThemeRef, mergedClsPrefixRef } = vue_1.inject(NotificationProvider_1.notificationProviderInjectionKey);
        return {
            mergedTheme: mergedThemeRef,
            mergedClsPrefix: mergedClsPrefixRef
        };
    },
    render() {
        const { $slots, scrollable, mergedClsPrefix, mergedTheme } = this;
        return (vue_1.h("div", { class: [
                `${mergedClsPrefix}-notification-container`,
                scrollable && `${mergedClsPrefix}-notification-container--scrollable`
            ] }, scrollable ? (vue_1.h(scrollbar_1.NScrollbar, { theme: mergedTheme.peers.Scrollbar, themeOverrides: mergedTheme.peerOverrides.Scrollbar }, $slots)) : ($slots)));
    }
});
