// naive-ui/lib/date-picker/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_scale_up_cssr_1 = require("../../../_styles/transitions/fade-in-scale-up.cssr"); // vars:
// --bezier
// --icon-color
// --icon-color-disabled
// --panel-border-radius
// --panel-color
// --panel-box-shadow
// --panel-text-color
// panel header
// --panel-header-padding
// --panel-header-divider-color
// panel calendar
// --calendar-left-padding
// --calendar-right-padding
// --calendar-title-height
// --calendar-title-padding
// --calendar-title-font-size
// --calendar-title-text-color
// --calendar-title-font-weight
// --calendar-title-grid-template-columns
// --calendar-days-height
// --calendar-days-divider-color
// --calendar-days-font-size
// --calendar-days-text-color
// --calendar-divider-color
// panel action
// --panel-action-padding
// --panel-action-divider-color
// panel item
// --item-border-radius
// --item-size
// --item-cell-width
// --item-cell-height
// --item-text-color
// --item-color-included
// --item-color-disabled
// --item-color-hover
// --item-color-active
// --item-font-size
// --item-text-color-disabled
// --item-text-color-active
// panel arrow
// --arrow-size
// --arrow-color


exports.default = cssr_1.c([cssr_1.cB('date-picker', `
 position: relative;
 z-index: auto;
 `, [cssr_1.cB('date-picker-icon', `
 color: var(--icon-color);
 transition: color .3s var(--bezier);
 `), cssr_1.cM('disabled', [cssr_1.cB('date-picker-icon', `
 color: var(--icon-color-disabled);
 `)])]), cssr_1.cB('date-panel', `
 outline: none;
 margin: 4px 0;
 display: grid;
 grid-template-columns: 0fr;
 border-radius: var(--panel-border-radius);
 background-color: var(--panel-color);
 box-shadow: var(--panel-box-shadow);
 color: var(--panel-text-color);
 `, [fade_in_scale_up_cssr_1.default(), cssr_1.cB('date-panel-calendar', {
  padding: 'var(--calendar-left-padding)',
  display: 'grid',
  gridTemplateColumns: '1fr',
  gridArea: 'left-calendar'
}, [cssr_1.cM('end', {
  padding: 'var(--calendar-right-padding)',
  gridArea: 'right-calendar'
})]), cssr_1.cM('date', {
  gridTemplateAreas: `
 "left-calendar"
 "footer"
 "action"
 `
}), cssr_1.cM('daterange', {
  gridTemplateAreas: `
 "left-calendar divider right-calendar"
 "footer footer footer"
 "action action action"
 `
}), cssr_1.cM('datetime', {
  gridTemplateAreas: `
 "header"
 "left-calendar"
 "footer"
 "action"
 `
}), cssr_1.cM('datetimerange', {
  gridTemplateAreas: `
 "header header header"
 "left-calendar divider right-calendar"
 "footer footer footer"
 "action action action"
 `
}), cssr_1.cB('date-panel-footer', {
  gridArea: 'footer'
}), cssr_1.cB('date-panel-actions', {
  gridArea: 'action'
}), cssr_1.cB('date-panel-header', {
  gridArea: 'header'
}), cssr_1.cB('date-panel-header', `
 box-sizing: border-box;
 width: 100%;
 align-items: center;
 padding: var(--panel-header-padding);
 display: flex;
 justify-content: space-between;
 border-bottom: 1px solid var(--panel-header-divider-color);
 `, [cssr_1.c('>', [cssr_1.c('*:not(:last-child)', {
  marginRight: '10px'
}), cssr_1.c('*', {
  flex: 1,
  width: 0
}), cssr_1.cB('time-picker', {
  zIndex: 1
})])]), cssr_1.cB('date-panel-month', `
 box-sizing: border-box;
 display: grid;
 grid-template-columns: var(--calendar-title-grid-template-columns);
 align-items: center;
 justify-items: center;
 padding: var(--calendar-title-padding);
 height: var(--calendar-title-height);
 `, [cssr_1.cE('prev, next, fast-prev, fast-next', `
 line-height: 0;
 cursor: pointer;
 width: var(--arrow-size);
 height: var(--arrow-size);
 color: var(--arrow-color);
 `), cssr_1.cE('month-year', `
 font-size: var(--calendar-title-font-size);
 font-weight: var(--calendar-title-font-weight);
 line-height: 17px;
 flex-grow: 1;
 text-align: center;
 color: var(--calendar-title-text-color);
 `)]), cssr_1.cB('date-panel-weekdays', `
 display: grid;
 margin: auto;
 grid-template-columns: repeat(7, var(--item-cell-width));
 grid-template-rows: repeat(1, var(--item-cell-height));
 align-items: center;
 justify-items: center;
 margin-bottom: 4px;
 border-bottom: 1px solid var(--calendar-days-divider-color);
 `, [cssr_1.cE('day', `
 line-height: 15px;
 width: var(--item-size);
 text-align: center;
 font-size: var(--calendar-days-font-size);
 color: var(--item-text-color);
 `)]), cssr_1.cB('date-panel-dates', `
 margin: auto;
 display: grid;
 grid-template-columns: repeat(7, var(--item-cell-width));
 grid-template-rows: repeat(6, var(--item-cell-height));
 align-items: center;
 justify-items: center;
 flex-wrap: wrap;
 `, [cssr_1.cB('date-panel-date', `
 position: relative;
 width: var(--item-size);
 height: var(--item-size);
 line-height: var(--item-size);
 text-align: center;
 font-size: var(--item-font-size);
 border-radius: var(--item-border-radius);
 z-index: 0;
 cursor: pointer;
 transition:
 background-color .2s var(--bezier),
 color .2s var(--bezier);
 `, [cssr_1.c('&.transition-disabled', {
  transition: 'none !important'
}, [cssr_1.c('&::before, &::after', {
  transition: 'none !important'
})]), cssr_1.cNotM('disabled', [cssr_1.cNotM('selected', [cssr_1.c('&:hover', {
  backgroundColor: 'var(--item-color-hover)'
})])]), cssr_1.cM('current', [cssr_1.cE('sup', `
 position: absolute;
 top: 2px;
 right: 2px;
 content: "";
 height: 4px;
 width: 4px;
 border-radius: 2px;
 background-color: var(--item-color-active);
 transition:
 background-color .2s var(--bezier);
 `)]), cssr_1.c('&::after', `
 content: "";
 z-index: -1;
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 border-radius: inherit;
 transition: background-color .3s var(--bezier);
 `), cssr_1.cM('covered, start, end', [cssr_1.cNotM('excluded', [cssr_1.c('&::before', `
 content: "";
 z-index: -2;
 position: absolute;
 left: calc((var(--item-size) - var(--item-cell-width)) / 2);
 right: calc((var(--item-size) - var(--item-cell-width)) / 2);
 top: 0;
 bottom: 0;
 background-color: var(--item-color-included);
 `), cssr_1.c('&:nth-child(7n + 1)::before', {
  borderTopLeftRadius: 'var(--item-border-radius)',
  borderBottomLeftRadius: 'var(--item-border-radius)'
}), cssr_1.c('&:nth-child(7n + 7)::before', {
  borderTopRightRadius: 'var(--item-border-radius)',
  borderBottomRightRadius: 'var(--item-border-radius)'
})])]), cssr_1.cM('selected', {
  color: 'var(--item-text-color-active)'
}, [cssr_1.c('&::after', {
  backgroundColor: 'var(--item-color-active)'
}), cssr_1.cM('start', [cssr_1.c('&::before', {
  left: '50%'
})]), cssr_1.cM('end', [cssr_1.c('&::before', {
  right: '50%'
})]), cssr_1.cE('sup', {
  backgroundColor: 'var(--panel-color)'
})]), cssr_1.cM('excluded', {
  color: 'var(--item-text-color-disabled)'
}, [cssr_1.cM('selected', [cssr_1.c('&::after', {
  backgroundColor: 'var(--item-color-disabled)'
})])]), cssr_1.cM('disabled', {
  cursor: 'not-allowed',
  color: 'var(--item-text-color-disabled)'
}, [cssr_1.cM('covered', [cssr_1.c('&::before', {
  backgroundColor: 'var(--item-color-disabled)'
})]), cssr_1.cM('selected', [cssr_1.c('&::before', {
  backgroundColor: 'var(--item-color-disabled)'
}), cssr_1.c('&::after', {
  backgroundColor: 'var(--item-color-disabled)'
})])])])]), cssr_1.cE('vertical-divider', `
 grid-area: divider;
 height: 100%;
 width: 1px;
 background-color: var(--calendar-divider-color);
 `), cssr_1.cB('date-panel-footer', {
  borderTop: '1px solid var(--panel-action-divider-color)',
  padding: 'var(--panel-extra-footer-padding)'
}), cssr_1.cB('date-panel-actions', `
 flex: 1;
 padding: var(--panel-action-padding);
 display: flex;
 align-items: center;
 justify-content: space-between;
 border-top: 1px solid var(--panel-action-divider-color);
 `, [cssr_1.cE('prefix, suffix', `
 display: flex;
 margin-bottom: -8px;
 `), cssr_1.cE('suffix', `
 align-self: flex-end;
 `), cssr_1.cE('prefix', `
 flex-wrap: wrap;
 `), cssr_1.cB('button', `
 margin-bottom: 8px;
 `, [cssr_1.c('&:not(:last-child)', `
 margin-right: 8px;
 `)])])])]);
