// naive-ui/lib/layout/src/styles/layout-sider.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --bezier
// --color
// --text-color
// --border-color
// --toggle-button-color
// --toggle-bar-color
// --toggle-bar-color-hover


exports.default = cssr_1.cB('layout-sider', `
 flex-shrink: 0;
 box-sizing: border-box;
 position: relative;
 z-index: 1;
 color: var(--text-color);
 transition:
 color .3s var(--bezier),
 border-color .3s var(--bezier),
 min-width .3s var(--bezier),
 max-width .3s var(--bezier),
 transform .3s var(--bezier),
 background-color .3s var(--bezier);
 background-color: var(--color);
 display: flex;
 justify-content: flex-end;
`, [cssr_1.cB('layout-toggle-button', `
 transition:
 transform .3s var(--bezier),
 fill .3s var(--bezier);
 cursor: pointer;
 width: 36px;
 height: 36px;
 position: absolute;
 top: 50%;
 right: 0;
 transform: translateX(50%) translateY(-50%);
 fill: var(--toggle-button-color);
 `), cssr_1.cB('layout-toggle-bar', `
 cursor: pointer;
 height: 72px;
 width: 32px;
 position: absolute;
 top: calc(50% - 36px);
 right: -28px;
 `, [cssr_1.cE('top, bottom', `
 position: absolute;
 width: 4px;
 border-radius: 2px;
 height: 38px;
 left: 14px;
 transition: 
 background-color .3s var(--bezier),
 transform .3s var(--bezier);
 `), cssr_1.cE('bottom', `
 position: absolute;
 top: 34px;
 `), cssr_1.c('&:hover', [cssr_1.cE('top', {
  transform: 'rotate(12deg) scale(1.15) translateY(-2px)'
}), cssr_1.cE('bottom', {
  transform: 'rotate(-12deg) scale(1.15) translateY(2px)'
})]), cssr_1.cM('collapsed', [cssr_1.c('&:hover', [cssr_1.cE('top', {
  transform: 'rotate(-12deg) scale(1.15) translateY(-2px)'
}), cssr_1.cE('bottom', {
  transform: 'rotate(12deg) scale(1.15) translateY(2px)'
})])]), cssr_1.cE('top, bottom', {
  backgroundColor: 'var(--toggle-bar-color)'
}), cssr_1.c('&:hover', [cssr_1.cE('top, bottom', {
  backgroundColor: 'var(--toggle-bar-color-hover)'
})])]), cssr_1.cE('border', `
 position: absolute;
 top: 0;
 right: 0;
 bottom: 0;
 width: 1px;
 transition: background-color .3s var(--bezier);
 `), cssr_1.cB('layout-sider-scroll-container', `
 flex-grow: 1;
 flex-shrink: 0;
 box-sizing: border-box;
 height: 100%;
 opacity: 0;
 transition: opacity .3s var(--bezier);
 max-width: 100%;
 `), cssr_1.cM('show-content', [cssr_1.cB('layout-sider-scroll-container', {
  opacity: 1
})]), cssr_1.cM('collapsed', [cssr_1.cB('layout-toggle-button', {
  transform: 'translateX(50%) translateY(-50%) rotate(180deg)'
})]), cssr_1.cM('absolute-positioned', `
 position: absolute;
 left: 0;
 top: 0;
 bottom: 0;
 `), cssr_1.cM('bordered', `
 border-right: 1px solid var(--border-color);
 `)]);
