// naive-ui/lib/timeline/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --bezier
// --circle-border
// --content-font-size
// --content-text-color
// --line-color
// --meta-text-color
// --title-font-size
// --title-font-weight
// --title-margin
// --title-text-color


exports.default = cssr_1.cB('timeline', `
 position: relative;
 width: 100%;
 display: flex;
 flex-direction: column;
 line-height: 1.25;
`, [cssr_1.cM('right-placement', [cssr_1.cB('timeline-item', [cssr_1.cB('timeline-item-content', `
 text-align: right;
 margin-right: 26px;
 `), cssr_1.cB('timeline-item-timeline', `
 width: 14px;
 right: 0;
 `)])]), cssr_1.cM('left-placement', [cssr_1.cB('timeline-item', [cssr_1.cB('timeline-item-content', `
 margin-left: 26px;
 `), cssr_1.cB('timeline-item-timeline', `
 left: 0;
 `)])]), cssr_1.cB('timeline-item', `
 position: relative;
 `, [cssr_1.c('&:last-child', [cssr_1.cB('timeline-item-timeline', [cssr_1.cE('line', `
 display: none;
 `)]), cssr_1.cB('timeline-item-content', [cssr_1.cE('meta', `
 margin-bottom: 0;
 `)])]), cssr_1.cB('timeline-item-content', [cssr_1.cE('title', `
 margin: var(--title-margin);
 font-size: var(--title-font-size);
 transition: color .3s var(--bezier);
 font-weight: var(--title-font-weight);
 margin-bottom: 6px;
 color: var(--title-text-color);
 `), cssr_1.cE('content', `
 transition: color .3s var(--bezier);
 font-size: var(--content-font-size);
 color: var(--content-text-color);
 `), cssr_1.cE('meta', `
 transition: color .3s var(--bezier);
 font-size: 12px;
 margin-top: 6px;
 margin-bottom: 20px;
 color: var(--meta-text-color);
 `)]), cssr_1.cB('timeline-item-timeline', `
 width: 26px;
 position: absolute;
 top: 2px;
 bottom: 0;
 height: 100%;
 `, [cssr_1.cE('circle', `
 border: var(--circle-border);
 transition:
 background-color .3s var(--bezier),
 border-color .3s var(--bezier);
 width: 14px;
 height: 14px;
 border-radius: 7px;
 box-sizing: border-box;
 `), cssr_1.cE('line', `
 transition: background-color .3s var(--bezier);
 position: absolute;
 top: 14px;
 left: 6px;
 bottom: 0px;
 width: 2px;
 background-color: var(--line-color);
 `)])])]);
