// naive-ui/lib/image/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_cssr_1 = require("../../../_styles/transitions/fade-in.cssr");

const fade_in_scale_up_cssr_1 = require("../../../_styles/transitions/fade-in-scale-up.cssr"); // vars:
// --icon-color
// --bezier


exports.default = cssr_1.c([cssr_1.c('body >', [cssr_1.cB('image-container', 'position: fixed;')]), cssr_1.cB('image-preview-container', `
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 display: flex;
 `), cssr_1.cB('image-preview-overlay', `
 z-index: -1;
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 background: rgba(0, 0, 0, .3);
 `, [fade_in_cssr_1.default()]), cssr_1.cB('image-preview-toolbar', `
 z-index: 1;
 position: absolute;
 left: 50%;
 transform: translateX(-50%);
 border-radius: 24px;
 height: 48px;
 bottom: 40px;
 padding: 0 12px;
 background: rgba(0, 0, 0, .35);
 color: var(--icon-color);
 transition: color .3s var(--bezier);
 display: flex;
 align-items: center;
 `, [cssr_1.cB('base-icon', `
 padding: 0 8px;
 font-size: 28px;
 cursor: pointer;
 `), fade_in_cssr_1.default()]), cssr_1.cB('image-preview-wrapper', `
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 display: flex;
 pointer-events: none;
 `, [fade_in_scale_up_cssr_1.default()]), cssr_1.cB('image-preview', `
 user-select: none;
 pointer-events: all;
 margin: auto;
 max-height: 100vh;
 max-width: 100vw;
 transition: transform .3s var(--bezier);
 `), cssr_1.cB('image', `
 display: inline-flex;
 cursor: pointer;
 `, [cssr_1.c('img', 'border-radius: inherit;')])]);
