// naive-ui/lib/_internal/select-menu/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../../_utils/cssr");

const fade_in_scale_up_cssr_1 = require("../../../../_styles/transitions/fade-in-scale-up.cssr"); // --loading-color
// --loading-size


exports.default = cssr_1.cB('base-select-menu', `
 outline: none;
 z-index: 0;
 position: relative;
 border-radius: var(--border-radius);
 transition:
 background-color .3s var(--bezier),
 box-shadow .3s var(--bezier);
 background-color: var(--color);
`, [cssr_1.cB('scrollbar', `
 max-height: var(--height);
 `), cssr_1.cB('virtual-list', `
 max-height: var(--height);
 `), cssr_1.cB('base-select-option', `
 height: var(--option-height);
 line-height: var(--option-height);
 font-size: var(--option-font-size);
 `), cssr_1.cB('base-select-group-header', `
 height: var(--option-height);
 line-height: var(--option-height);
 font-size: .93em;
 `), cssr_1.cB('base-select-menu-option-wrapper', `
 position: relative;
 width: 100%;
 `), cssr_1.cE('loading, empty', `
 display: flex;
 padding: 12px 32px;
 flex: 1;
 justify-content: center;
 `), cssr_1.cE('loading', `
 color: var(--loading-color);
 font-size: var(--loading-size);
 `), cssr_1.cE('action', `
 padding: 8px var(--option-padding-left);
 font-size: var(--option-font-size);
 transition: 
 color .3s var(--bezier);
 border-color .3s var(--bezier);
 border-top: 1px solid var(--action-divider-color);
 color: var(--action-text-color);
 `), cssr_1.cB('base-select-group-header', `
 position: relative;
 cursor: default;
 padding: var(--option-padding);
 color: var(--group-header-text-color);
 `), cssr_1.cB('base-select-option', `
 cursor: pointer;
 position: relative;
 padding: var(--option-padding);
 white-space: nowrap;
 transition:
 background-color .3s var(--bezier),
 color .3s var(--bezier),
 opacity .3s var(--bezier);
 text-overflow: ellipsis;
 overflow: hidden;
 box-sizing: border-box;
 color: var(--option-text-color);
 opacity: 1;
 `, [cssr_1.c('&:active', {
  color: 'var(--option-text-color-pressed)'
}), cssr_1.cM('grouped', {
  paddingLeft: 'calc(var(--option-padding-left) * 1.5)'
}), cssr_1.cM('selected', {
  color: 'var(--option-text-color-active)'
}), cssr_1.cM('disabled', {
  cursor: 'not-allowed'
}, [cssr_1.cNotM('selected', {
  color: 'var(--option-text-color-disabled)'
}), cssr_1.cM('selected', {
  opacity: 'var(--option-opacity-disabled)'
})]), cssr_1.cM('pending', {
  backgroundColor: 'var(--option-color-pending)'
}), cssr_1.cE('check', `
 font-size: 14px;
 position: absolute;
 right: 8px;
 top: calc(var(--option-height) / 2 - 7px);
 color: var(--option-check-color);
 transition: color .3s var(--bezier);
 `, [fade_in_scale_up_cssr_1.default({
  enterScale: '0.5'
})])]), cssr_1.cM('multiple', [cssr_1.cB('base-select-option', `
 position: relative;
 padding-right: 28px;
 `)])]);
