// naive-ui/lib/button/src/styles/button.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_width_expand_cssr_1 = require("../../../_styles/transitions/fade-in-width-expand.cssr");

const icon_switch_cssr_1 = require("../../../_styles/transitions/icon-switch.cssr"); // vars:
// --bezier
// --ripple-duration
// --opacity-disabled
// --text-color
// --text-color-hover
// --text-color-pressed
// --text-color-focus
// --text-color-disabled
// --color
// --color-hover
// --color-pressed
// --color-focus
// --color-disabled
// --border
// --border-hover
// --border-pressed
// --border-focus
// --border-disabled
// --ripple-color
// --border-radius
// --height
// --width
// --font-size
// --padding
// --icon-size
// --icon-margin
// --wave-opacity
// --font-weight
//
// private-vars:
// --border-color-xxx, used for custom color


exports.default = cssr_1.c([cssr_1.cB('button', `
 font-weight: var(--font-weight);
 line-height: 1;
 font-family: inherit;
 padding: var(--padding);
 height: var(--height);
 font-size: var(--font-size);
 border-radius: var(--border-radius);
 color: var(--text-color);
 background-color: var(--color);
 width: var(--width);
 white-space: nowrap;
 outline: none;
 position: relative;
 z-index: auto;
 border: none;
 display: inline-flex;
 flex-wrap: nowrap;
 align-items: center;
 justify-content: center;
 user-select: none;
 text-align: center;
 cursor: pointer;
 text-decoration: none;
 transition:
 color .3s var(--bezier),
 background-color .3s var(--bezier),
 opacity .3s var(--bezier),
 border-color .3s var(--bezier);
 `, [cssr_1.cM('color', [cssr_1.cE('border', {
  borderColor: 'var(--border-color)'
}), cssr_1.cM('disabled', [cssr_1.cE('border', {
  borderColor: 'var(--border-color-disabled)'
})]), cssr_1.cNotM('disabled', [cssr_1.c('&:focus', [cssr_1.cE('state-border', {
  borderColor: 'var(--border-color-focus)'
})]), cssr_1.c('&:hover', [cssr_1.cE('state-border', {
  borderColor: 'var(--border-color-hover)'
})]), cssr_1.c('&:active', [cssr_1.cE('state-border', {
  borderColor: 'var(--border-color-pressed)'
})]), cssr_1.cM('pressed', [cssr_1.cE('state-border', {
  borderColor: 'var(--border-color-pressed)'
})])])]), cssr_1.cM('disabled', {
  backgroundColor: 'var(--color-disabled)',
  color: 'var(--text-color-disabled)'
}, [cssr_1.cE('border', {
  border: 'var(--border-disabled)'
})]), cssr_1.cNotM('disabled', [cssr_1.c('&:focus', {
  backgroundColor: 'var(--color-focus)',
  color: 'var(--text-color-focus)'
}, [cssr_1.cE('state-border', {
  border: 'var(--border-focus)'
})]), cssr_1.c('&:hover', {
  backgroundColor: 'var(--color-hover)',
  color: 'var(--text-color-hover)'
}, [cssr_1.cE('state-border', {
  border: 'var(--border-hover)'
})]), cssr_1.c('&:active', {
  backgroundColor: 'var(--color-pressed)',
  color: 'var(--text-color-pressed)'
}, [cssr_1.cE('state-border', {
  border: 'var(--border-pressed)'
})]), cssr_1.cM('pressed', {
  backgroundColor: 'var(--color-pressed)',
  color: 'var(--text-color-pressed)'
}, [cssr_1.cE('state-border', {
  border: 'var(--border-pressed)'
})])]), cssr_1.cB('base-wave', `
 pointer-events: none;
 top: 0;
 right: 0;
 bottom: 0;
 left: 0;
 animation-iteration-count: 1;
 animation-duration: var(--ripple-duration);
 animation-timing-function: var(--bezier-ease-out), var(--bezier-ease-out);
 `, [cssr_1.cM('active', {
  zIndex: 1,
  animationName: 'button-wave-spread, button-wave-opacity'
})]), typeof window !== 'undefined' && 'MozBoxSizing' in document.createElement('div').style ? cssr_1.c('&::moz-focus-inner', {
  border: 0
}) : null, cssr_1.cE('border, state-border', `
 position: absolute;
 left: 0;
 top: 0;
 right: 0;
 bottom: 0;
 border-radius: inherit;
 transition: border-color .3s var(--bezier);
 pointer-events: none;
 `), cssr_1.cE('border', {
  border: 'var(--border)'
}), cssr_1.cE('state-border', {
  border: 'var(--border)',
  borderColor: '#0000',
  zIndex: 1
}), cssr_1.cE('icon', `
 margin: var(--icon-margin);
 margin-left: 0;
 height: var(--icon-size);
 width: var(--icon-size);
 max-width: var(--icon-size);
 font-size: var(--icon-size);
 position: relative;
 flex-shrink: 0;
 `, [cssr_1.cB('icon-slot', `
 height: var(--icon-size);
 width: var(--icon-size);
 position: absolute;
 left: 0;
 top: 50%;
 transform: translateY(-50%);
 display: flex;
 `, [icon_switch_cssr_1.default({
  top: '50%',
  originalTransform: 'translateY(-50%)'
})]), fade_in_width_expand_cssr_1.default()]), cssr_1.cE('content', `
 display: flex;
 align-items: center;
 flex-wrap: nowrap;
 `, [cssr_1.c('~', [cssr_1.cE('icon', {
  margin: 'var(--icon-margin)',
  marginRight: 0
})])]), cssr_1.cM('block', `
 display: flex;
 width: 100%;
 `), cssr_1.cM('dashed', [cssr_1.cE('border, state-border', {
  borderStyle: 'dashed !important'
})]), cssr_1.cM('disabled', {
  cursor: 'not-allowed',
  opacity: 'var(--opacity-disabled)'
})]), cssr_1.c('@keyframes button-wave-spread', {
  from: {
    boxShadow: '0 0 0.5px 0 var(--ripple-color)'
  },
  to: {
    // don't use exact 5px since chrome will display the animation with glitches
    boxShadow: '0 0 0.5px 4.5px var(--ripple-color)'
  }
}), cssr_1.c('@keyframes button-wave-opacity', {
  from: {
    opacity: 'var(--wave-opacity)'
  },
  to: {
    opacity: 0
  }
})]);
