// naive-ui/lib/_internal/suffix/src/Suffix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const clear_1 = require("../../clear");
const loading_1 = require("../../loading");
const icon_1 = require("../../icon");
const icons_1 = require("../../icons");
exports.default = vue_1.defineComponent({
    name: 'InternalSelectionSuffix',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        showArrow: {
            type: Boolean,
            default: undefined
        },
        showClear: {
            type: Boolean,
            default: undefined
        },
        loading: Boolean,
        onClear: Function
    },
    setup(props) {
        return () => {
            const { clsPrefix } = props;
            return (vue_1.h(loading_1.default, { clsPrefix: clsPrefix, class: `${clsPrefix}-base-suffix`, strokeWidth: 24, scale: 0.85, show: props.loading }, {
                default: () => props.showArrow ? (vue_1.h(clear_1.default, { clsPrefix: clsPrefix, show: props.showClear, onClear: props.onClear }, {
                    default: () => (vue_1.h(icon_1.default, { clsPrefix: clsPrefix, class: `${clsPrefix}-base-suffix__arrow` }, { default: () => vue_1.h(icons_1.ChevronDownIcon, null) }))
                })) : null
            }));
        };
    }
});
