// naive-ui/lib/transfer/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_cssr_1 = require("../../../_styles/transitions/fade-in.cssr");

const animation = cssr_1.c([cssr_1.c('@keyframes transfer-slide-in-from-left', `
 0% {
 transform: translateX(-150%);
 }
 100% {
 transform: translateX(0);
 }
 `), cssr_1.c('@keyframes transfer-slide-out-to-right', `
 0% {
 transform: translateX(0);
 }
 100% {
 transform: translateX(150%);
 }
 `), cssr_1.c('@keyframes transfer-slide-in-from-right', `
 0% {
 transform: translateX(150%);
 }
 100% {
 transform: translateX(0);
 }
 `), cssr_1.c('@keyframes transfer-slide-out-to-left', `
 0% {
 transform: translateX(0);
 }
 100% {
 transform: translateX(-150%);
 }
 `), cssr_1.c('@keyframes transfer-height-collapse', `
 0% {
 max-height: var(--item-height);
 }
 100% {
 max-height: 0;
 }
 `), cssr_1.c('@keyframes transfer-height-expand', `
 0% {
 max-height: 0;
 }
 100% {
 max-height: var(--item-height);
 }
 `)]);
exports.default = cssr_1.c([cssr_1.cB('transfer', `
 display: flex;
 width: var(--width);
 font-size: var(--font-size);
 `, [cssr_1.cM('filterable', [cssr_1.cB('transfer-list', [cssr_1.cB('transfer-list-body', {
  height: 'calc(var(--item-height) * 5.6 + 45px)'
})])]), cssr_1.cB('transfer-icon', `
 color: var(--icon-color);
 transition: color .3s var(--bezier);
 `), cssr_1.cM('disabled', [cssr_1.cB('transfer-icon', {
  color: 'var(--icon-color-disabled)'
})]), cssr_1.cB('transfer-list', `
 background-clip: padding-box;
 width: calc(50% - 36px);
 position: relative;
 transition: background-color .3s var(--bezier);
 border-radius: var(--border-radius);
 background-color: var(--list-color);
 `, [cssr_1.cE('border', `
 border: 1px solid var(--border-color);
 transition: border-color .3s var(--bezier);
 pointer-events: none;
 border-radius: inherit;
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 `), cssr_1.cB('transfer-list-header', `
 height: calc(var(--item-height) + 4px);
 box-sizing: border-box;
 display: flex;
 align-items: center;
 background-clip: padding-box;
 border-radius: inherit;
 border-bottom-left-radius: 0;
 border-bottom-right-radius: 0;
 background-color: var(--header-color);
 transition:
 border-color .3s var(--bezier),
 background-color .3s var(--bezier);
 `, [cssr_1.cE('checkbox', `
 display: flex;
 align-items: center;
 position: relative;
 padding: 0 9px 0 14px;
 `), cssr_1.cE('header', `
 flex: 1;
 line-height: 1;
 font-weight: var(--header-font-weight);
 transition: color .3s var(--bezier);
 color: var(--header-text-color);
 `, [cssr_1.cM('disabled', {
  color: 'var(--header-text-color-disabled)'
})]), cssr_1.cE('extra', `
 transition: color .3s var(--bezier);
 font-size: var(--extra-font-size);
 justify-self: flex-end;
 margin-right: 14px;
 white-space: nowrap;
 color: var(--header-extra-text-color);
 `)]), cssr_1.cB('transfer-list-body', `
 height: calc(var(--item-height) * 5.6);
 box-sizing: border-box;
 overflow: hidden;
 position: relative;
 display: flex;
 flex-direction: column;
 border-radius: inherit;
 border-top-left-radius: 0;
 border-top-right-radius: 0;
 `, [cssr_1.cB('transfer-filter', `
 padding: 0 8px 8px 8px;
 box-sizing: border-box;
 background-color: var(--header-color);
 transition:
 border-color .3s var(--bezier),
 background-color .3s var(--bezier);
 border-bottom: 1px solid var(--filter-divider-color);
 `), cssr_1.cB('transfer-list-flex-container', `
 flex: 1;
 position: relative;
 `, [cssr_1.cB('scrollbar', `
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 height: unset;
 `, [cssr_1.cB('scrollbar-content', {
  width: '100%'
})]), cssr_1.cB('empty', `
 position: absolute;
 left: 50%;
 top: 50%;
 transform: translateY(-50%) translateX(-50%);
 `, [fade_in_cssr_1.default()]), cssr_1.cB('transfer-list-content', `
 padding: 0;
 margin: 0;
 position: relative;
 `, [cssr_1.cM('transition-disabled', [cssr_1.cB('transfer-list-item', {
  animation: 'none !important'
})]), cssr_1.cB('transfer-list-item', `
 height: var(--item-height);
 max-height: var(--item-height);
 transition:
 background-color .3s var(--bezier),
 color .3s var(--bezier);
 position: relative;
 cursor: pointer;
 display: flex;
 align-items: center;
 color: var(--item-text-color);
 `, [cssr_1.cNotM('disabled', [cssr_1.c('&:hover', {
  backgroundColor: 'var(--item-color-pending)'
})]), cssr_1.cE('extra', `
 text-overflow: ellipsis;
 overflow: hidden;
 white-space: nowrap;
 padding-right: 4px;
 `), cssr_1.cE('checkbox', `
 display: flex;
 align-items: center;
 position: relative;
 padding: 0 9px 0 14px;
 `), cssr_1.cM('disabled', `
 cursor: not-allowed
 background-color: #0000;
 color: var(--item-text-color-disabled);
 `), cssr_1.cM('source', {
  animationFillMode: 'forwards'
}, [cssr_1.c('&.item-enter-active', `
 transform: translateX(150%);
 animation-duration: .25s, .25s;
 animation-timing-function: var(--bezier), var(--bezier-ease-out);
 animation-delay: 0s, .25s;
 animation-name: transfer-height-expand, transfer-slide-in-from-right;
 `), cssr_1.c('&.item-leave-active', `
 transform: translateX(-150%);
 animation-duration: .25s, .25s;
 animation-timing-function: var(--bezier), var(--bezier-ease-in);
 animation-delay: .25s, 0s;
 animation-name: transfer-height-collapse, transfer-slide-out-to-right;
 `)]), cssr_1.cM('target', {
  animationFillMode: 'forwards'
}, [cssr_1.c('&.item-enter-active', `
 transform: translateX(-150%);
 animation-duration: .25s, .25s;
 animation-timing-function: var(--bezier), var(--bezier-ease-out);
 animation-delay: 0s, .25s;
 animation-name: transfer-height-expand, transfer-slide-in-from-left;
 `), cssr_1.c('&.item-leave-active', `
 transform: translateX(150%);
 animation-duration: .25s, .25s;
 animation-timing-function: var(--bezier), var(--bezier-ease-in);
 animation-delay: .25s, 0s;
 animation-name: transfer-height-collapse, transfer-slide-out-to-left;
 `)])])])])])]), cssr_1.cB('transfer-gap', {
  width: '72px',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  flexDirection: 'column'
}), cssr_1.cB('button', [cssr_1.c('&:first-child', {
  marginBottom: '12px'
})])]), animation]);
