// naive-ui/lib/notification/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --color
// --text-color
// --description-text-color
// --action-text-color
// --title-text-color
// --title-font-weight
// --bezier
// --bezier-ease-out
// --bezier-ease-in
// --border-radius
// --box-shadow
// --close-margin
// --close-size
// --close-color
// --close-color-hover
// --close-color-pressed
// --line-height
// --icon-color
// --width
// --padding-top
// --padding-bottom
// --padding-left
// --padding-right


exports.default = cssr_1.c([cssr_1.cB('notification-container', `
 z-index: 4000;
 position: fixed;
 top: 12px;
 left: 0;
 right: 0;
 height: 0;
 overflow: visible;
 display: flex;
 flex-direction: column;
 align-items: flex-end;
 `, [cssr_1.c('>', [cssr_1.cB('scrollbar', `
 width: initial;
 overflow: visible;
 height: -moz-fit-content !important;
 height: fit-content !important;
 max-height: 100vh !important;
 `, [cssr_1.c('>', [cssr_1.cB('scrollbar-container', `
 height: -moz-fit-content !important;
 height: fit-content !important;
 max-height: 100vh !important;
 `, [cssr_1.cB('scrollbar-content', `
 padding-top: 12px;
 padding-bottom: 33px;
 `)])])])]), cssr_1.cM('scrollable', {
  top: 0
}), cssr_1.cB('notification', [cssr_1.c('&.notification-transition-enter-from, &.notification-transition-leave-to', `
 opacity: 0;
 margin-bottom: 0 !important;
 transform: translateX(calc(100% + 16px));
 `), cssr_1.c('&.notification-transition-leave-from, &.notification-transition-enter-to', `
 opacity: 1;
 transform: translateX(0);
 `), cssr_1.c('&.notification-transition-leave-active', `
 transition:
 background-color .3s var(--bezier),
 color .3s var(--bezier),
 opacity .3s var(--bezier),
 transform .3s var(--bezier-ease-in),
 max-height .3s var(--bezier),
 margin-bottom .3s linear,
 box-shadow .3s var(--bezier);
 `)]), cssr_1.cB('notification', `
 background-color: var(--color);
 color: var(--text-color);
 transition:
 background-color .3s var(--bezier),
 color .3s var(--bezier),
 opacity .3s var(--bezier),
 transform .3s var(--bezier-ease-out),
 max-height .3s var(--bezier),
 margin-bottom .3s linear,
 box-shadow .3s var(--bezier);
 font-family: inherit;
 font-size: var(--font-size);
 font-weight: 400;
 position: relative;
 display: flex;
 overflow: hidden;
 flex-shrink: 0;
 margin-bottom: 12px;
 margin-left: 28px;
 margin-right: 16px;
 padding-left: var(--padding-left);
 padding-right: var(--padding-right);
 width: var(--width);
 border-radius: var(--border-radius);
 box-shadow: var(--box-shadow);
 box-sizing: border-box;
 opacity: 1;
 `, [// TODO: refactor type styles & transition
cssr_1.cE('avatar', [cssr_1.cB('icon', {
  color: 'var(--icon-color)'
}), cssr_1.cB('base-icon', {
  color: 'var(--icon-color)'
})]), cssr_1.cM('show-avatar', [cssr_1.cB('notification-main', `
 margin-left: 40px;
 width: calc(100% - 40px); 
 `)]), cssr_1.cM('closable', [cssr_1.cB('notification-main', [cssr_1.c('> *:first-child', {
  paddingRight: '20px'
})]), cssr_1.cE('close', `
 position: absolute;
 top: 0;
 right: 0;
 margin: var(--close-margin);
 font-size: var(--close-size);
 transition: color .3s var(--bezier);
 `)]), cssr_1.cE('avatar', `
 position: absolute;
 top: var(--padding-top);
 left: var(--padding-left);
 width: 28px;
 height: 28px;
 font-size: 28px;
 `, [cssr_1.cB('icon', {
  transition: 'color .3s var(--bezier)'
})]), cssr_1.cB('notification-main', `
 padding-top: var(--padding-top);
 padding-bottom: var(--padding-bottom);
 box-sizing: border-box;
 display: flex;
 flex-direction: column;
 margin-left: 8px;
 width: calc(100% - 8px);
 `, [cssr_1.cB('notification-main-footer', `
 display: flex;
 align-items: center;
 justify-content: space-between;
 margin-top: 12px;
 `, [cssr_1.cE('meta', `
 font-size: 12px;
 transition: color .3s var(--bezier-ease-out);
 color: var(--description-text-color);
 `), cssr_1.cE('action', `
 cursor: pointer;
 transition: color .3s var(--bezier-ease-out);
 color: var(--action-text-color);
 `)]), cssr_1.cE('header', `
 font-weight: var(--title-font-weight);
 font-size: 16px;
 transition: color .3s var(--bezier-ease-out);
 color: var(--title-text-color);
 `), cssr_1.cE('description', `
 margin-top: 8px;
 font-size: 12px;
 transition: color .3s var(--bezier-ease-out);
 color: var(--description-text-color);
 `), cssr_1.cE('content', `
 line-height: var(--line-height);
 margin: 12px 0 0 0;
 font-family: inherit;
 white-space: pre-wrap;
 word-wrap: break-word;
 transition: color .3s var(--bezier-ease-out);
 color: var(--text-color);
 `, [cssr_1.c('&:first-child', {
  margin: 0
})])])])])]);
