// naive-ui/lib/form/src/FormItemGridItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formItemGiPropKeys = exports.formItemGiProps = void 0;
const vue_1 = require("vue");
const GridItem_1 = require("../../grid/src/GridItem");
const _utils_1 = require("../../_utils");
const FormItem_1 = require("./FormItem");
exports.formItemGiProps = Object.assign(Object.assign({}, GridItem_1.gridItemProps), FormItem_1.formItemProps);
exports.formItemGiPropKeys = _utils_1.keysOf(exports.formItemGiProps);
exports.default = vue_1.defineComponent({
    __GRID_ITEM__: true,
    name: 'FormItemGridItem',
    alias: ['FormItemGi'],
    props: exports.formItemGiProps,
    setup(props) {
        const formItemInstRef = vue_1.ref(null);
        const validate = ((...args) => {
            const { value } = formItemInstRef;
            if (value) {
                return value.validate(...args);
            }
        });
        const restoreValidation = () => {
            const { value } = formItemInstRef;
            if (value) {
                return value.restoreValidation();
            }
        };
        return {
            validate,
            restoreValidation
        };
    },
    render() {
        return vue_1.h(GridItem_1.default, _utils_1.keep(this.$props, GridItem_1.gridItemPropKeys), {
            default: () => {
                const itemProps = _utils_1.keep(this.$props, FormItem_1.formItemPropKeys);
                return vue_1.h(FormItem_1.default, Object.assign({ ref: 'formItemInstRef' }, itemProps), this.$slots);
            }
        });
    }
});
