// naive-ui/lib/data-table/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fixedColumnStyle = createFixedColumnStyle(); // vars:
// --font-size
// --th-padding
// --td-padding
// --bezier
// --border-radius
// --line-height
// --border-color
// --border-color-modal
// --border-color-popover
// --th-color
// --th-color-hover
// --th-color-modal
// --th-color-hover-modal
// --th-color-popover
// --th-color-hover-popover
// --td-color
// --td-color-hover
// --td-color-modal
// --td-color-hover-modal
// --td-color-popover
// --td-color-hover-popover
// --th-text-color
// --td-text-color
// --th-font-weight
// --th-button-color-hover
// --th-icon-color
// --th-icon-color-active
// --filter-size
// --action-divider-color
// --action-padding
// --action-button-margin
// --pagination-margin
// --empty-padding
// --sorter-size
// --box-shadow-before used in Body.tsx
// --box-shadow-after used in Body.tsx

exports.default = cssr_1.c([cssr_1.cB('data-table', `
 width: 100%;
 font-size: var(--font-size);
 --merged-th-color: var(--th-color);
 --merged-td-color: var(--td-color);
 --merged-border-color: var(--border-color);
 --merged-th-color-hover: var(--th-color-hover);
 --merged-td-color-hover: var(--td-color-hover);
 `, [cssr_1.cB('data-table-expand-trigger', 'cursor: pointer;'), cssr_1.cB('data-table-expand-placeholder', `
 margin-right: 8px;
 display: inline-block;
 width: 16px;
 height: 1px;
 `), cssr_1.cB('data-table-indent', `
 display: inline-block;
 height: 1px;
 `), cssr_1.cB('data-table-expand-trigger', `
 margin-right: 8px;
 cursor: pointer;
 font-size: 16px;
 vertical-align: -.125em;
 `), cssr_1.cB('data-table-thead', {
  transition: 'background-color .3s var(--bezier)',
  backgroundColor: 'var(--merged-th-color)'
}), cssr_1.cB('data-table-tr', {
  boxSizing: 'border-box',
  backgroundClip: 'padding-box',
  transition: 'background-color .3s var(--bezier)'
}, [cssr_1.c('&:hover', {
  backgroundColor: 'var(--merged-td-color-hover)'
}, [cssr_1.cB('data-table-td', {
  backgroundColor: 'var(--merged-td-color-hover)'
})])]), cssr_1.cB('data-table-th', `
 padding: var(--th-padding);
 position: relative;
 text-align: start;
 box-sizing: border-box;
 background-color: var(--merged-th-color);
 border-color: var(--merged-border-color);
 border-bottom: 1px solid var(--merged-border-color);
 color: var(--th-text-color);
 transition:
 border-color .3s var(--bezier),
 color .3s var(--bezier),
 background-color .3s var(--bezier);
 font-weight: var(--th-font-weight);
 `, [cssr_1.cM('filterable', {
  paddingRight: '36px'
}), fixedColumnStyle, cssr_1.cM('selection', `
 padding: 0;
 text-align: center;
 line-height: 0;
 z-index: 3;
 `), cssr_1.cE('ellipsis', `
 display: inline-block;
 vertical-align: bottom;
 text-overflow: ellipsis;
 overflow: hidden;
 white-space: nowrap;
 max-width: 100%;
 `), cssr_1.cM('hover', {
  backgroundColor: 'var(--merged-th-color-hover)'
}), cssr_1.cM('sortable', {
  cursor: 'pointer'
}, [cssr_1.cE('ellipsis', {
  maxWidth: 'calc(100% - 18px)'
}), cssr_1.c('&:hover', {
  backgroundColor: 'var(--merged-th-color-hover)'
})]), cssr_1.cB('data-table-sorter', `
 height: var(--sorter-size);
 width: var(--sorter-size);
 margin-left: 4px;
 position: relative;
 display: inline-flex;
 vertical-align: -0.2em;
 color: var(--th-icon-color);
 transition: 
 transform .3s var(--bezier),
 color .3s var(--bezier);
 `, [cssr_1.cM('desc', {
  transform: 'rotate(0)'
}), cssr_1.cM('asc', {
  transform: 'rotate(-180deg)'
}), cssr_1.cM('asc, desc', {
  color: 'var(--th-icon-color-active)'
})]), cssr_1.cB('data-table-filter', `
 position: absolute;
 z-index: auto;
 right: 0;
 width: 36px;
 top: 0;
 bottom: 0;
 cursor: pointer;
 display: flex;
 justify-content: center;
 align-items: center;
 transition:
 background-color .3s var(--bezier),
 color .3s var(--bezier);
 font-size: var(--filter-size);
 color: var(--th-icon-color);
 `, [cssr_1.c('&:hover', `
 background-color: var(--th-button-color-hover);
 `), cssr_1.cM('show', `
 background-color: var(--th-button-color-hover);
 `), cssr_1.cM('active', `
 background-color: var(--th-button-color-hover);
 color: var(--th-icon-color-active);
 `)])]), cssr_1.cB('data-table-td', `
 padding: var(--td-padding);
 text-align: start;
 box-sizing: border-box;
 border: none;
 background-color: var(--merged-td-color);
 color: var(--td-text-color);
 border-bottom: 1px solid var(--merged-border-color);
 transition:
 box-shadow .3s var(--bezier),
 background-color .3s var(--bezier),
 border-color .3s var(--bezier),
 color .3s var(--bezier);
 `, [cssr_1.cM('last-row', {
  borderBottom: '0 solid var(--merged-border-color)'
}, [// make sure there is no overlap between bottom border and
// fixed column box shadow
cssr_1.c('&::after', {
  bottom: '0 !important'
}), cssr_1.c('&::before', {
  bottom: '0 !important'
})]), cssr_1.cM('summary', `
 background-color: var(--merged-th-color);
 `), cssr_1.cM('hover', {
  backgroundColor: 'var(--merged-td-color-hover)'
}), cssr_1.cM('ellipsis', `
 text-overflow: ellipsis;
 overflow: hidden;
 white-space: nowrap;
 `), cssr_1.cM('selection, expand', `
 text-align: center;
 padding: 0;
 line-height: 0;
 `), fixedColumnStyle]), cssr_1.cB('data-table-empty', `
 padding: var(--empty-padding);
 flex-grow: 1;
 flex-shrink: 0;
 opacity: 1;
 display: flex;
 align-items: center;
 justify-content: center;
 transition: opacity .3s var(--bezier);
 `, [cssr_1.cM('hide', {
  opacity: 0
})]), cssr_1.cE('pagination', `
 margin: var(--pagination-margin);
 display: flex;
 justify-content: flex-end;
 `), cssr_1.cB('data-table-wrapper', `
 position: relative;
 transition: border-color .3s var(--bezier);
 border-top-left-radius: var(--border-radius);
 border-top-right-radius: var(--border-radius);
 line-height: var(--line-height);
 `), cssr_1.cM('single-column', [cssr_1.cB('data-table-td', {
  borderBottom: '0 solid var(--merged-border-color)'
}, [cssr_1.c('&::after, &::before', {
  bottom: '0 !important'
})])]), cssr_1.cNotM('single-line', [cssr_1.cB('data-table-th', {
  borderRight: '1px solid var(--merged-border-color)'
}, [cssr_1.cM('last', {
  borderRight: '0 solid var(--merged-border-color)'
})]), cssr_1.cB('data-table-td', {
  borderRight: '1px solid var(--merged-border-color)'
}, [cssr_1.cM('last-col', {
  borderRight: '0 solid var(--merged-border-color)'
})])]), cssr_1.cM('bordered', [cssr_1.cB('data-table-wrapper', {
  border: '1px solid var(--merged-border-color)',
  borderBottomLeftRadius: 'var(--border-radius)',
  borderBottomRightRadius: 'var(--border-radius)'
}), cssr_1.cB('data-table-td', [cssr_1.cM('last-row', {
  borderBottom: '0 solid var(--merged-border-color)'
})]), cssr_1.cB('data-table-base-table-body', `
 border-bottom-left-radius: calc(var(--border-radius) - 1px);
 border-bottom-right-radius: calc(var(--border-radius) - 1px);
 `)]), cssr_1.cB('data-table-base-table', [cssr_1.cM('transition-disabled', [cssr_1.cB('data-table-th', [cssr_1.c('&::after, &::before', {
  transition: 'none'
})]), cssr_1.cB('data-table-td', [cssr_1.c('&::after, &::before', {
  transition: 'none'
})])])]), cssr_1.cM('bottom-bordered', [cssr_1.cB('data-table-td', [cssr_1.cM('last-row', {
  borderBottom: '1px solid var(--merged-border-color)'
})])]), cssr_1.cB('data-table-table', `
 font-variant-numeric: tabular-nums;
 width: 100%;
 word-wrap: break-word;
 word-break: break-all;
 transition: background-color .3s var(--bezier);
 border-collapse: separate;
 border-spacing: 0;
 background-color: var(--merged-td-color)
 `), cssr_1.cB('data-table-base-table-header', `
 border-top-left-radius: calc(var(--border-radius) - 1px);
 border-top-right-radius: calc(var(--border-radius) - 1px);
 z-index: 3;
 overflow: scroll;
 flex-shrink: 0;
 transition: border-color .3s var(--bezier);
 scrollbar-width: none;
 `, [cssr_1.c('&::-webkit-scrollbar', {
  width: 0,
  height: 0
})]), cssr_1.cB('data-table-check-extra', `
 transition: color .3s var(--bezier);
 color: var(--th-icon-color);
 position: absolute;
 font-size: 14px;
 right: -4px;
 top: 50%;
 transform: translateY(-50%);
 z-index: 1;
 `)]), cssr_1.cB('data-table-filter-menu', [cssr_1.cB('scrollbar', {
  maxHeight: '240px'
}), cssr_1.cE('group', {
  display: 'flex',
  flexDirection: 'column',
  padding: '12px 12px 0 12px'
}, [cssr_1.cB('checkbox', {
  marginBottom: '12px',
  marginRight: 0
}), cssr_1.cB('radio', {
  marginBottom: '12px',
  marginRight: 0
})]), cssr_1.cE('action', `
 padding: var(--action-padding);
 display: flex;
 flex-wrap: nowrap;
 justify-content: space-evenly;
 border-top: 1px solid var(--action-divider-color);
 `, [cssr_1.cB('button', [cssr_1.c('&:not(:last-child)', {
  margin: 'var(--action-button-margin)'
}), cssr_1.c('&:last-child', {
  marginRight: 0
})])]), cssr_1.cB('divider', {
  margin: '0!important'
})]), cssr_1.insideModal(cssr_1.cB('data-table', `
 --merged-th-color: var(--th-color-modal);
 --merged-td-color: var(--td-color-modal);
 --merged-border-color: var(--border-color-modal);
 --merged-th-color-hover: var(--th-color-hover-modal);
 --merged-td-color-hover: var(--td-color-hover-modal);
 `)), cssr_1.insidePopover(cssr_1.cB('data-table', `
 --merged-th-color: var(--th-color-popover);
 --merged-td-color: var(--td-color-popover);
 --merged-border-color: var(--border-color-popover);
 --merged-th-color-hover: var(--th-color-hover-popover);
 --merged-td-color-hover: var(--td-color-hover-popover);
 `))]);

function createFixedColumnStyle() {
  return [cssr_1.cM('fixed-left', `
 left: 0;
 position: sticky;
 z-index: 2;
 `, [cssr_1.c('&::after', `
 pointer-events: none;
 content: "";
 width: 36px;
 display: inline-block;
 position: absolute;
 top: 0;
 bottom: -1px;
 transition: box-shadow .2s var(--bezier);
 right: -36px;
 `)]), cssr_1.cM('fixed-right', {
    right: 0,
    position: 'sticky',
    zIndex: 1
  }, [cssr_1.c('&::before', `
 pointer-events: none;
 content: "";
 width: 36px;
 display: inline-block;
 position: absolute;
 top: 0;
 bottom: -1px;
 transition: box-shadow .2s var(--bezier);
 left: -36px;
 `)])];
}
