// naive-ui/lib/spin/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_cssr_1 = require("../../../_styles/transitions/fade-in.cssr"); // vars:
// --bezier
// --opacity-spinning
// --size
// --color


exports.default = cssr_1.c([cssr_1.c('@keyframes spin-rotate', `
 from {
 transform: rotate(0);
 }
 to {
 transform: rotate(360deg);
 }
 `), cssr_1.cB('spin-container', {
  position: 'relative'
}, [cssr_1.cB('spin', `
 position: absolute;
 top: 50%;
 left: 50%;
 transform: translateX(-50%) translateY(-50%);
 `, [fade_in_cssr_1.default()]), cssr_1.cB('spin-content', `
 opacity: 1;
 transition: opacity .3s var(--bezier);
 pointer-events: all;
 `, [cssr_1.cM('spinning', {
  pointerEvents: 'none',
  opacity: 'var(--opacity-spinning)'
})])]), cssr_1.cB('spin', `
 display: inline-flex;
 height: var(--size);
 width: var(--size);
 font-size: var(--size);
 color: var(--color);
 `, [cssr_1.cM('rotate', `
 animation: spin-rotate 2s linear infinite;
 `)])]);
