// naive-ui/lib/carousel/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --bezier
// --dot-color
// --dot-color-active
// --dot-size


exports.default = cssr_1.cB('carousel', `
 overflow: hidden;
 position: relative;
`, [cssr_1.cE('slides', `
 transition: transform .3s var(--bezier);
 display: flex;
 `, [cssr_1.c('> div', `
 overflow: hidden;
 `, [cssr_1.c('> img', `
 display: block;
 `)])]), cssr_1.cE('dots', `
 position: absolute;
 display: flex;
 flex-wrap: nowrap;
 `), cssr_1.cE('dot', `
 height: var(--dot-size);
 width: var(--dot-size);
 background-color: var(--dot-color);
 border-radius: 50%;
 cursor: pointer;
 transition:
 box-shadow .3s var(--bezier),
 background-color .3s var(--bezier);
 outline: none;
 `, [cssr_1.c('&:focus', `
 background-color: var(--dot-color-active);
 `), cssr_1.cM('active', `
 background-color: var(--dot-color-active);
 `), cssr_1.c('&:last-child', `
 margin-right: 0;
 `)]), cssr_1.cM('left', [cssr_1.cE('slides', `
 flex-direction: column;
 `), cssr_1.cE('dots', `
 transform: translateY(-50%);
 top: 50%;
 left: 16px;
 flex-direction: column;
 `), cssr_1.cE('dot', `
 margin-bottom: 12px;
 `)]), cssr_1.cM('right', [cssr_1.cE('slides', `
 flex-direction: column;
 `), cssr_1.cE('dots', `
 transform: translateY(-50%);
 top: 50%;
 right: 16px;
 flex-direction: column;
 `), cssr_1.cE('dot', `
 margin-bottom: 12px;
 `)]), cssr_1.cM('top', [cssr_1.cE('dots', `
 transform: translateX(-50%);
 top: 16px;
 left: 50%;
 `), cssr_1.cE('dot', `
 margin-right: 12px;
 `)]), cssr_1.cM('bottom', [cssr_1.cE('dots', `
 transform: translateX(-50%);
 bottom: 16px;
 left: 50%;
 `), cssr_1.cE('dot', `
 margin-right: 12px;
 `)])]);
