// naive-ui/lib/steps/src/Steps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stepsInjectionKey = void 0;
const vue_1 = require("vue");
const _mixins_1 = require("../../_mixins");
const styles_1 = require("../styles");
const index_cssr_1 = require("./styles/index.cssr");
const _utils_1 = require("../../_utils");
function stepWithIndex(step, i) {
    if (typeof step !== 'object' || step === null || Array.isArray(step)) {
        return null;
    }
    if (!step.props)
        step.props = {};
    step.props.internalIndex = i + 1;
    return step;
}
function stepsWithIndex(steps) {
    return steps.map((step, i) => stepWithIndex(step, i));
}
const stepsProps = Object.assign(Object.assign({}, _mixins_1.useTheme.props), { current: Number, status: {
        type: String,
        default: 'process'
    }, size: {
        type: String,
        default: 'medium'
    }, vertical: Boolean });
exports.stepsInjectionKey = Symbol('steps');
exports.default = vue_1.defineComponent({
    name: 'Steps',
    props: stepsProps,
    setup(props, { slots }) {
        const { mergedClsPrefixRef } = _mixins_1.useConfig(props);
        const themeRef = _mixins_1.useTheme('Steps', 'Steps', index_cssr_1.default, styles_1.stepsLight, props, mergedClsPrefixRef);
        vue_1.provide(exports.stepsInjectionKey, {
            props,
            mergedThemeRef: themeRef,
            mergedClsPrefixRef,
            stepsSlots: slots
        });
        return {
            mergedClsPrefix: mergedClsPrefixRef
        };
    },
    render() {
        const { mergedClsPrefix } = this;
        return (vue_1.h("div", { class: [
                `${mergedClsPrefix}-steps`,
                this.vertical && `${mergedClsPrefix}-steps--vertical`
            ] }, stepsWithIndex(_utils_1.getSlot(this))));
    }
});
