// naive-ui/lib/timeline/src/Timeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.timelineInjectionKey = void 0;
const vue_1 = require("vue");
const _mixins_1 = require("../../_mixins");
const styles_1 = require("../styles");
const index_cssr_1 = require("./styles/index.cssr");
const timelineProps = Object.assign(Object.assign({}, _mixins_1.useTheme.props), { itemPlacement: {
        type: String,
        default: 'left'
    }, size: {
        type: String,
        default: 'medium'
    } });
exports.timelineInjectionKey = Symbol('timeline');
exports.default = vue_1.defineComponent({
    name: 'Timeline',
    props: timelineProps,
    setup(props, { slots }) {
        const { mergedClsPrefixRef } = _mixins_1.useConfig(props);
        const themeRef = _mixins_1.useTheme('Timeline', 'Timeline', index_cssr_1.default, styles_1.timelineLight, props, mergedClsPrefixRef);
        vue_1.provide(exports.timelineInjectionKey, {
            props,
            mergedThemeRef: themeRef,
            mergedClsPrefixRef
        });
        return () => {
            const { value: mergedClsPrefix } = mergedClsPrefixRef;
            return (vue_1.h("div", { class: [
                    `${mergedClsPrefix}-timeline`,
                    `${mergedClsPrefix}-timeline--${props.size}-size`,
                    `${mergedClsPrefix}-timeline--${props.itemPlacement}-placement`
                ] }, slots));
        };
    }
});
