// naive-ui/lib/data-table/src/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNextSorter = exports.isColumnFilterable = exports.isColumnSortable = exports.shouldUseArrayInSingleMode = exports.createRowClassName = exports.createCustomWidthStyle = exports.getFlagOfOrder = exports.createShallowClonedObject = exports.getColKey = exports.getColWidth = exports.expandColWidth = exports.selectionColWidth = void 0;
const seemly_1 = require("seemly");
exports.selectionColWidth = 40;
exports.expandColWidth = 40;
function getColWidth(col) {
    if (col.type === 'selection')
        return exports.selectionColWidth;
    if (col.type === 'expand')
        return exports.expandColWidth;
    if ('children' in col)
        return undefined;
    return col.width;
}
exports.getColWidth = getColWidth;
function getColKey(col) {
    if (col.type === 'selection')
        return '__n_selection__';
    if (col.type === 'expand')
        return '__n_expand__';
    return col.key;
}
exports.getColKey = getColKey;
function createShallowClonedObject(object) {
    if (!object)
        return object;
    if (typeof object === 'object') {
        return Object.assign({}, object);
    }
    return object;
}
exports.createShallowClonedObject = createShallowClonedObject;
function getFlagOfOrder(order) {
    if (order === 'ascend')
        return 1;
    else if (order === 'descend')
        return -1;
    return 0;
}
exports.getFlagOfOrder = getFlagOfOrder;
function createCustomWidthStyle(column) {
    return {
        width: seemly_1.pxfy(getColWidth(column))
    };
}
exports.createCustomWidthStyle = createCustomWidthStyle;
function createRowClassName(row, index, rowClassName) {
    if (typeof rowClassName === 'function')
        return rowClassName(row, index);
    return rowClassName || '';
}
exports.createRowClassName = createRowClassName;
// for compatibility
// If column.filterOptionValues or column.defaultFilterOptionValues is set, use
// array value
function shouldUseArrayInSingleMode(column) {
    return (column.filterOptionValues !== undefined ||
        (column.filterOptionValue === undefined &&
            column.defaultFilterOptionValues !== undefined));
}
exports.shouldUseArrayInSingleMode = shouldUseArrayInSingleMode;
function isColumnSortable(column) {
    if ('children' in column)
        return false;
    return !!column.sorter;
}
exports.isColumnSortable = isColumnSortable;
function isColumnFilterable(column) {
    if ('children' in column)
        return false;
    return (!!column.filter && (!!column.filterOptions || !!column.renderFilterMenu));
}
exports.isColumnFilterable = isColumnFilterable;
function getNextOrderOf(order) {
    if (!order)
        return 'descend';
    else if (order === 'descend')
        return 'ascend';
    return false;
}
function createNextSorter(column, currentSortState) {
    if (column.sorter === undefined)
        return null;
    if (currentSortState === null || currentSortState.columnKey !== column.key) {
        return {
            columnKey: column.key,
            sorter: column.sorter,
            order: getNextOrderOf(false)
        };
    }
    else {
        return Object.assign(Object.assign({}, currentSortState), { order: getNextOrderOf(currentSortState.order) });
    }
}
exports.createNextSorter = createNextSorter;
