// naive-ui/lib/log/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_scale_up_cssr_1 = require("../../../_styles/transitions/fade-in-scale-up.cssr"); // vars:
// --bezier
// --loading-color
// --loader-border
// --loader-color
// --loader-text-color
// --loader-font-size
// --loading-color


exports.default = cssr_1.cB('log', `
 position: relative;
 box-sizing: border-box;
 transition:
 border-color .3s var(--bezier),
 color .3s var(--bezier);
`, [cssr_1.cE('lines', `
 margin: 0;
 white-space: pre-wrap;
 `), cssr_1.c('pre', `
 margin: 0;
 `), cssr_1.cB('log-loader', `
 transition:
 color .3s var(--bezier),
 background-color .3s var(--bezier),
 border-color .3s var(--bezier);
 box-sizing: border-box;
 position: absolute;
 right: 16px;
 top: 8px;
 height: 34px;
 border-radius: 17px;
 line-height: 34px;
 white-space: nowrap;
 overflow: hidden;
 border: var(--loader-border);
 color: var(--loader-text-color);
 background-color: var(--loader-color);
 font-size: var(--loader-font-size);
 `, [fade_in_scale_up_cssr_1.default(), cssr_1.cE('content', `
 display: inline-block;
 vertical-align: bottom;
 line-height: 34px;
 padding-left: 40px;
 padding-right: 20px;
 white-space: nowrap;
 `), cssr_1.cB('base-loading', `
 color: var(--loading-color);
 position: absolute;
 left: 12px;
 top: calc(50% - 10px);
 font-size: 20px;
 width: 20px;
 height: 20px;
 display: inline-block;
 `)])]);
