// naive-ui/lib/_internal/clear/src/Clear.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _mixins_1 = require("../../../_mixins");
const icons_1 = require("../../icons");
const icon_1 = require("../../icon");
const icon_switch_transition_1 = require("../../icon-switch-transition");
const index_cssr_1 = require("./styles/index.cssr");
exports.default = vue_1.defineComponent({
    name: 'BaseClear',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        show: {
            type: Boolean,
            default: false
        },
        onClear: Function
    },
    setup(props) {
        _mixins_1.useStyle('BaseClear', index_cssr_1.default, vue_1.toRef(props, 'clsPrefix'));
        const { NConfigProvider } = _mixins_1.useConfig();
        return {
            NConfigProvider,
            handleMouseDown(e) {
                e.preventDefault();
            }
        };
    },
    render() {
        const { clsPrefix } = this;
        return (vue_1.h("div", { class: `${clsPrefix}-base-clear` },
            vue_1.h(icon_switch_transition_1.default, null, {
                default: () => {
                    return this.show ? (vue_1.h(icon_1.default, { clsPrefix: clsPrefix, key: "dismiss", class: `${clsPrefix}-base-clear__clear`, onClick: this.onClear, onMousedown: this.handleMouseDown, "data-clear": true }, {
                        default: () => vue_1.h(icons_1.ClearIcon, null)
                    })) : (vue_1.h("div", { key: "icon", class: `${clsPrefix}-base-clear__placeholder` }, this.$slots));
                }
            })));
    }
});
