// naive-ui/lib/breadcrumb/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --font-size
// --bezier
// --item-text-color
// --item-text-color-hover
// --item-text-color-pressed
// --item-text-color-active
// --separator-color
// --font-weight-active


exports.default = cssr_1.cB('breadcrumb', `
 white-space: nowrap;
`, [cssr_1.cB('breadcrumb-item', {
  fontSize: 'var(--font-size)',
  transition: 'color .3s var(--bezier)'
}, [cssr_1.cB('icon', `
 font-size: 18px;
 vertical-align: -.2em;
 transition: color .3s var(--bezier);
 color: var(--item-text-color);
 `), cssr_1.cE('link', {
  cursor: 'pointer',
  transition: 'color .3s var(--bezier)',
  color: 'var(--item-text-color)'
}), cssr_1.cE('separator', {
  margin: '0 8px',
  color: 'var(--separator-color)',
  transition: 'color .3s var(--bezier)'
}), cssr_1.c('&:hover', [cssr_1.cB('icon', {
  color: 'var(--item-text-color-hover)'
}), cssr_1.cE('link', {
  color: 'var(--item-text-color-hover)'
})]), cssr_1.c('&:active', [cssr_1.cB('icon', {
  color: 'var(--item-text-color-pressed)'
}), cssr_1.cE('link', {
  color: 'var(--item-text-color-pressed)'
})]), cssr_1.c('&:last-child', [cssr_1.cE('link', `
 font-weight: var(--font-weight-active);
 cursor: unset;
 color: var(--item-text-color-active);
 `), cssr_1.cB('icon', {
  color: 'var(--item-text-color-active)'
}), cssr_1.cE('separator', {
  display: 'none'
})])])]);
