// vueuc/lib/lazy-teleport/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const vooks_1 = require("vooks");
const v_node_1 = require("../../shared/v-node");
exports.default = vue_1.defineComponent({
    name: 'LazyTeleport',
    props: {
        to: {
            type: [String, Object],
            default: undefined
        },
        disabled: Boolean,
        show: {
            type: Boolean,
            required: true
        }
    },
    setup(props) {
        return {
            showTeleport: vooks_1.useFalseUntilTruthy(vue_1.toRef(props, 'show')),
            mergedTo: vue_1.computed(() => {
                const { to } = props;
                return to !== null && to !== void 0 ? to : 'body';
            })
        };
    },
    render() {
        return this.showTeleport
            ? this.disabled
                ? v_node_1.getSlot(this.$slots)
                : vue_1.h(vue_1.Teleport, {
                    disabled: this.disabled,
                    to: this.mergedTo
                }, v_node_1.getSlot(this.$slots))
            : null;
    }
});
