// naive-ui/es/time-picker/src/PanelCol.js
import { h, defineComponent } from 'vue';
export default defineComponent({
    name: 'TimePickerPanelCol',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        data: {
            type: Array,
            required: true
        },
        activeValue: {
            type: Number,
            default: null
        },
        onItemClick: Function
    },
    render() {
        const { activeValue, onItemClick, clsPrefix } = this;
        return this.data.map((item) => {
            const { value, disabled } = item;
            const numValue = Number(value);
            const active = activeValue === numValue;
            return (h("div", { key: value, "data-active": active ? '' : null, class: [
                    `${clsPrefix}-time-picker-col__item`,
                    {
                        [`${clsPrefix}-time-picker-col__item--active`]: active,
                        [`${clsPrefix}-time-picker-col__item--disabled`]: disabled
                    }
                ], onClick: onItemClick && !disabled ? () => onItemClick(numValue) : undefined }, value));
        });
    }
});
