// naive-ui/lib/back-top/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_scale_up_cssr_1 = require("../../../_styles/transitions/fade-in-scale-up.cssr"); // vars:
// --bezier
// --border-radius
// --height
// --width
// --box-shadow
// --box-shadow-hover
// --box-shadow-pressed
// --color
// --icon-size
// --icon-color
// --icon-color-hover
// --icon-color-pressed
// --text-color


exports.default = cssr_1.cB('back-top', `
 position: fixed;
 right: 40px;
 bottom: 40px;
 cursor: pointer;
 display: flex;
 align-items: center;
 justify-content: center;
 color: var(--text-color);
 transition:
 color .3s var(--bezier),
 box-shadow .3s var(--bezier),
 background-color .3s var(--bezier);
 border-radius: var(--border-radius);
 height: var(--height);
 min-width: var(--width);
 box-shadow: var(--box-shadow);
 background-color: var(--color);
`, [fade_in_scale_up_cssr_1.default(), cssr_1.cM('transition-disabled', {
  transition: 'none !important'
}), cssr_1.cB('base-icon', `
 font-size: var(--icon-size);
 color: var(--icon-color);
 transition: color .3s var(--bezier);
 `), cssr_1.c('svg', {
  pointerEvents: 'none'
}), cssr_1.c('&:hover', {
  boxShadow: 'var(--box-shadow-hover)'
}, [cssr_1.cB('base-icon', {
  color: 'var(--icon-color-hover)'
})]), cssr_1.c('&:active', {
  boxShadow: 'var(--box-shadow-pressed)'
}, [cssr_1.cB('base-icon', {
  color: 'var(--icon-color-pressed)'
})])]);
