// naive-ui/lib/_mixins/use-form-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formItemInjectionKey = void 0;
const vue_1 = require("vue");
exports.formItemInjectionKey = Symbol('formItem');
function useFormItem(props, { defaultSize = 'medium', mergedSize } = {}) {
    const NFormItem = vue_1.inject(exports.formItemInjectionKey, null);
    vue_1.provide(exports.formItemInjectionKey, null);
    const mergedSizeRef = vue_1.computed(mergedSize
        ? () => mergedSize(NFormItem)
        : () => {
            const { size } = props;
            if (size)
                return size;
            if (NFormItem) {
                const { mergedSize } = NFormItem;
                if (mergedSize.value !== undefined) {
                    return mergedSize.value;
                }
            }
            return defaultSize;
        });
    vue_1.onBeforeUnmount(() => {
        if (NFormItem) {
            NFormItem.restoreValidation();
        }
    });
    return {
        mergedSizeRef,
        nTriggerFormBlur() {
            if (NFormItem) {
                NFormItem.handleContentBlur();
            }
        },
        nTriggerFormChange() {
            if (NFormItem) {
                NFormItem.handleContentChange();
            }
        },
        nTriggerFormFocus() {
            if (NFormItem) {
                NFormItem.handleContentFocus();
            }
        },
        nTriggerFormInput() {
            if (NFormItem) {
                NFormItem.handleContentInput();
            }
        }
    };
}
exports.default = useFormItem;
