// naive-ui/lib/modal/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_cssr_1 = require("../../../_styles/transitions/fade-in.cssr");

const fade_in_scale_up_cssr_1 = require("../../../_styles/transitions/fade-in-scale-up.cssr"); // vars:
// --bezier-ease-out
// --box-shadow
// --color
// --text-color


exports.default = cssr_1.c([cssr_1.cB('modal-container', `
 position: fixed;
 left: 0;
 top: 0;
 height: 0;
 width: 0;
 display: flex;
 `), cssr_1.cB('modal-mask', `
 position: fixed;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 background-color: rgba(0, 0, 0, .4);
 `, [fade_in_cssr_1.default({
  enterDuration: '.25s',
  leaveDuration: '.25s',
  enterCubicBezier: 'var(--bezier-ease-out)',
  leaveCubicBezier: 'var(--bezier-ease-out)'
})]), cssr_1.cB('modal-body-wrapper', `
 position: fixed;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 overflow: visible;
 `, [cssr_1.cB('modal-scroll-content', `
 min-height: 100%;
 display: flex;
 position: relative;
 `)]), cssr_1.cB('modal', `
 align-self: center;
 color: var(--text-color);
 margin: auto;
 box-shadow: var(--box-shadow);
 `, [fade_in_scale_up_cssr_1.default({
  duration: '.25s',
  enterScale: '.5'
})])]);
