// naive-ui/lib/dialog/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --icon-color
// --bezier
// --close-margin
// --icon-margin
// --icon-size
// --close-size
// --close-color
// --close-color-hover
// --close-color-pressed
// --color
// --text-color
// --border-radius
// --padding
// --line-height
// --border
// --content-margin
// --title-font-size
// --title-font-weight
// --title-text-color
// --action-space


exports.default = cssr_1.c([cssr_1.cB('dialog', `
 line-height: var(--line-height);
 position: relative;
 background: var(--color);
 color: var(--text-color);
 box-sizing: border-box;
 margin: auto;
 border-radius: var(--border-radius);
 padding: var(--padding);
 transition: 
 border-color .3s var(--bezier),
 background-color .3s var(--bezier),
 color .3s var(--bezier);
 `, [cssr_1.cE('icon', {
  color: 'var(--icon-color)'
}), cssr_1.cM('bordered', {
  border: 'var(--border)'
}), cssr_1.cM('icon-top', [cssr_1.cE('close', {
  margin: 'var(--close-margin)'
}), cssr_1.cE('icon', {
  margin: 'var(--icon-margin)'
}), cssr_1.cE('content', {
  textAlign: 'center'
}), cssr_1.cE('title', {
  justifyContent: 'center'
}), cssr_1.cE('action', {
  justifyContent: 'center'
})]), cssr_1.cM('icon-left', [cssr_1.cE('icon', {
  margin: 'var(--icon-margin)'
})]), cssr_1.cE('close', `
 font-size: var(--close-size);
 position: absolute;
 right: 0;
 top: 0;
 margin: var(--close-margin);
 transition: .3s color var(--bezier);
 `), cssr_1.cE('content', `
 font-size: var(--font-size);
 margin: var(--content-margin);
 position: relative;
 `), cssr_1.cE('action', `
 display: flex;
 justify-content: flex-end;
 `, [cssr_1.c('> *:not(:last-child)', {
  marginRight: 'var(--action-space)'
})]), cssr_1.cE('icon', {
  fontSize: 'var(--icon-size)',
  transition: 'color .3s var(--bezier)'
}), cssr_1.cE('title', `
 transition: color .3s var(--bezier);
 display: flex;
 align-items: center;
 font-size: var(--title-font-size);
 font-weight: var(--title-font-weight);
 color: var(--title-text-color);
 `), cssr_1.cB('dialog-icon-container', {
  display: 'flex',
  justifyContent: 'center'
})]), cssr_1.insideModal(cssr_1.cB('dialog', `
 width: 446px;
 max-width: calc(100vw - 32px);
 `)), cssr_1.cB('dialog', [cssr_1.asModal(`
 width: 446px;
 max-width: calc(100vw - 32px);
 `)])]);
