// naive-ui/lib/button/src/ButtonGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buttonGroupInjectionKey = void 0;
const vue_1 = require("vue");
const _mixins_1 = require("../../_mixins");
const button_group_cssr_1 = require("./styles/button-group.cssr");
exports.buttonGroupInjectionKey = Symbol('button-group');
const buttonGroupProps = {
    size: {
        type: String,
        default: undefined
    },
    vertical: Boolean
};
exports.default = vue_1.defineComponent({
    name: 'ButtonGroup',
    props: buttonGroupProps,
    setup(props) {
        const { mergedClsPrefixRef } = _mixins_1.useConfig(props);
        _mixins_1.useStyle('ButtonGroup', button_group_cssr_1.default, mergedClsPrefixRef);
        vue_1.provide(exports.buttonGroupInjectionKey, props);
        return {
            mergedClsPrefix: mergedClsPrefixRef
        };
    },
    render() {
        const { mergedClsPrefix } = this;
        return (vue_1.h("div", { class: [
                `${mergedClsPrefix}-button-group`,
                this.vertical && `${mergedClsPrefix}-button-group--vertical`
            ], role: "group" }, this.$slots));
    }
});
