// naive-ui/lib/time-picker/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_scale_up_cssr_1 = require("../../../_styles/transitions/fade-in-scale-up.cssr"); // vars:
// --icon-color
// --icon-color-disabled
// --bezier
// --border-radius
// --item-color-hover
// --item-font-size
// --item-height
// --item-opacity-disabled
// --item-text-color
// --item-text-color-active
// --item-width
// --panel-action-padding
// --panel-box-shadow
// --panel-color
// --panel-divider-color


exports.default = cssr_1.c([cssr_1.cB('time-picker', `
 z-index: auto;
 position: relative;
 `, [cssr_1.cB('time-picker-icon', `
 color: var(--icon-color);
 transition: color .3s var(--bezier);
 `), cssr_1.cM('disabled', [cssr_1.cB('time-picker-icon', {
  color: 'var(--icon-color-disabled)'
})])]), cssr_1.cB('time-picker-panel', `
 outline: none;
 font-size: var(--item-font-size);
 border-radius: var(--border-radius);
 margin: 4px 0;
 min-width: 104px;
 overflow: hidden;
 background-color: var(--panel-color);
 box-shadow: var(--panel-box-shadow);
 `, [fade_in_scale_up_cssr_1.default(), cssr_1.cB('time-picker-actions', `
 padding: var(--panel-action-padding);
 align-items: center;
 display: flex;
 justify-content: space-evenly;
 `), cssr_1.cB('time-picker-cols', `
 height: calc(var(--item-height) * 7);
 display: flex;
 position: relative;
 border-bottom: 1px solid var(--panel-divider-color);
 `), cssr_1.cB('time-picker-col', `
 flex-grow: 1;
 min-width: var(--item-width);
 height: calc(var(--item-height) * 7);
 flex-direction: column;
 transition: box-shadow .3s var(--bezier);
 `, [cssr_1.cM('transition-disabled', [cssr_1.cE('item', {
  transition: 'none'
})]), cssr_1.cE('padding', {
  height: 'calc(var(--item-height) * 6)'
}), cssr_1.cE('item', `
 cursor: pointer;
 height: var(--item-height);
 display: flex;
 align-items: center;
 justify-content: center;
 transition: 
 color .3s var(--bezier),
 background-color .3s var(--bezier),
 opacity .3s var(--bezier),
 text-decoration-color .3s var(--bezier);
 background: #0000;
 text-decoration-color: #0000;
 color: var(--item-text-color);
 `, [cssr_1.cNotM('disabled', [cssr_1.c('&:hover', {
  backgroundColor: 'var(--item-color-hover)'
})]), cssr_1.cM('active', `
 background-color: var(--item-color-hover);
 color: var(--item-text-color-active);
 `), cssr_1.cM('disabled', `
 opacity: var(--item-opacity-disabled);
 cursor: not-allowed;
 `)]), cssr_1.cM('invalid', [cssr_1.cE('item', [cssr_1.cM('active', `
 text-decoration: line-through;
 text-decoration-color: var(--item-text-color-active);
 `)])])])])]);
