// naive-ui/lib/table/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NTbody = exports.NThead = exports.NTd = exports.NTr = exports.NTh = exports.NTable = void 0;
var Table_1 = require("./src/Table");
Object.defineProperty(exports, "NTable", { enumerable: true, get: function () { return Table_1.default; } });
var Th_1 = require("./src/Th");
Object.defineProperty(exports, "NTh", { enumerable: true, get: function () { return Th_1.default; } });
var Tr_1 = require("./src/Tr");
Object.defineProperty(exports, "NTr", { enumerable: true, get: function () { return Tr_1.default; } });
var Td_1 = require("./src/Td");
Object.defineProperty(exports, "NTd", { enumerable: true, get: function () { return Td_1.default; } });
var Thead_1 = require("./src/Thead");
Object.defineProperty(exports, "NThead", { enumerable: true, get: function () { return Thead_1.default; } });
var Tbody_1 = require("./src/Tbody");
Object.defineProperty(exports, "NTbody", { enumerable: true, get: function () { return Tbody_1.default; } });
