// naive-ui/lib/layout/styles/dark.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("../../_styles/common");
const styles_1 = require("../../scrollbar/styles");
const seemly_1 = require("seemly");
const layoutDark = {
    name: 'Layout',
    common: common_1.commonDark,
    peers: {
        Scrollbar: styles_1.scrollbarDark
    },
    self(vars) {
        const { textColor2, bodyColor, cardColor, dividerColor, scrollbarColor, scrollbarColorHover } = vars;
        return {
            textColor: textColor2,
            textColorInverted: textColor2,
            color: bodyColor,
            colorEmbedded: bodyColor,
            headerColor: cardColor,
            headerColorInverted: cardColor,
            footerColor: cardColor,
            footerColorInverted: cardColor,
            headerBorderColor: dividerColor,
            headerBorderColorInverted: dividerColor,
            footerBorderColor: dividerColor,
            footerBorderColorInverted: dividerColor,
            siderBorderColor: dividerColor,
            siderBorderColorInverted: dividerColor,
            siderColor: cardColor,
            siderColorInverted: cardColor,
            siderToggleButtonColor: 'rgba(255, 255, 255, .3)',
            siderToggleBarColor: seemly_1.composite(bodyColor, scrollbarColor),
            siderToggleBarColorHover: seemly_1.composite(bodyColor, scrollbarColorHover),
            __invertScrollbar: 'false'
        };
    }
};
exports.default = layoutDark;
