// naive-ui/lib/time/src/Time.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const date_fns_1 = require("date-fns");
const _mixins_1 = require("../../_mixins");
const timeProps = {
    time: {
        type: [Number, Date],
        default: () => Date.now()
    },
    type: {
        type: String,
        default: 'datetime'
    },
    to: {
        type: [Number, Date],
        default: () => Date.now()
    },
    unix: {
        type: Boolean,
        default: false
    },
    format: String,
    text: {
        type: Boolean,
        default: false
    }
};
exports.default = vue_1.defineComponent({
    name: 'Time',
    props: timeProps,
    setup(props) {
        const { localeRef, dateLocaleRef } = _mixins_1.useLocale('Time');
        const dateFnsOptionsRef = vue_1.computed(() => {
            return {
                locale: dateLocaleRef.value.locale
            };
        });
        const mergedTimeRef = vue_1.computed(() => {
            if (props.unix) {
                return date_fns_1.fromUnixTime(props.time);
            }
            return props.time;
        });
        const mergedToRef = vue_1.computed(() => {
            if (props.unix) {
                return date_fns_1.fromUnixTime(props.to);
            }
            return props.to;
        });
        const renderedTimeRef = vue_1.computed(() => {
            if (props.format) {
                return date_fns_1.format(mergedTimeRef.value, props.format, dateFnsOptionsRef.value);
            }
            else if (props.type === 'date') {
                return date_fns_1.format(mergedTimeRef.value, localeRef.value.dateFormat, dateFnsOptionsRef.value);
            }
            else if (props.type === 'datetime') {
                return date_fns_1.format(mergedTimeRef.value, localeRef.value.dateTimeFormat, dateFnsOptionsRef.value);
            }
            else {
                return date_fns_1.formatDistance(mergedTimeRef.value, mergedToRef.value, {
                    addSuffix: true,
                    locale: dateLocaleRef.value.locale
                });
            }
        });
        return {
            renderedTime: renderedTimeRef
        };
    },
    render() {
        return this.text
            ? vue_1.createTextVNode(this.renderedTime)
            : vue_1.h('time', [this.renderedTime]);
    }
});
