// naive-ui/lib/_internal/close/src/Close.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _mixins_1 = require("../../../_mixins");
const icon_1 = require("../../icon");
const icons_1 = require("../../icons");
const index_cssr_1 = require("./styles/index.cssr");
exports.default = vue_1.defineComponent({
    name: 'BaseClose',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        disabled: {
            type: Boolean,
            default: undefined
        },
        onClick: Function
    },
    setup(props) {
        _mixins_1.useStyle('BaseClose', index_cssr_1.default, vue_1.toRef(props, 'clsPrefix'));
        return () => {
            const { clsPrefix, disabled } = props;
            return (vue_1.h(icon_1.default, { role: "button", ariaDisabled: disabled, ariaLabel: "close", clsPrefix: clsPrefix, class: [
                    `${clsPrefix}-base-close`,
                    disabled && `${clsPrefix}-base-close--disabled`
                ], onClick: disabled ? undefined : props.onClick }, {
                default: () => vue_1.h(icons_1.CloseIcon, null)
            }));
        };
    }
});
