// naive-ui/es/dropdown/src/DropdownGroupHeader.js
import { defineComponent, h, inject } from 'vue';
import { render } from '../../_utils';
import { dropdownMenuInjectionKey } from './DropdownMenu';
export default defineComponent({
    name: 'DropdownGroupHeader',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        tmNode: {
            type: Object,
            required: true
        }
    },
    setup() {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const { showIconRef, hasSubmenuRef } = inject(dropdownMenuInjectionKey);
        return {
            showIcon: showIconRef,
            hasSubmenu: hasSubmenuRef
        };
    },
    render() {
        var _a;
        const { clsPrefix, hasSubmenu, showIcon } = this;
        const { rawNode } = this.tmNode;
        return (h("div", { class: `${clsPrefix}-dropdown-option` },
            h("div", { class: `${clsPrefix}-dropdown-option-body ${clsPrefix}-dropdown-option-body--group` },
                h("div", { "__dropdown-option": true, class: [
                        `${clsPrefix}-dropdown-option-body__prefix`,
                        showIcon && `${clsPrefix}-dropdown-option-body__prefix--show-icon`
                    ] }, render(rawNode.icon)),
                h("div", { class: `${clsPrefix}-dropdown-option-body__label`, "__dropdown-option": true }, render((_a = rawNode.label) !== null && _a !== void 0 ? _a : rawNode.title)),
                h("div", { class: [
                        `${clsPrefix}-dropdown-option-body__suffix`,
                        hasSubmenu &&
                            `${clsPrefix}-dropdown-option-body__suffix--has-submenu`
                    ], "__dropdown-option": true }))));
    }
});
