// naive-ui/lib/image/src/Image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const ImagePreview_1 = require("./ImagePreview");
const ImageGroup_1 = require("./ImageGroup");
const _mixins_1 = require("../../_mixins");
const imageProps = {
    alt: String,
    height: [String, Number],
    imgProps: Object,
    width: [String, Number],
    src: String,
    showToolbar: { type: Boolean, default: true }
};
exports.default = vue_1.defineComponent({
    name: 'Image',
    props: imageProps,
    inheritAttrs: false,
    setup(props) {
        const imageRef = vue_1.ref(null);
        const imgPropsRef = vue_1.toRef(props, 'imgProps');
        const previewInstRef = vue_1.ref(null);
        const imageGroupHandle = vue_1.inject(ImageGroup_1.imageGroupInjectionKey, null);
        const { mergedClsPrefixRef } = imageGroupHandle || _mixins_1.useConfig(props);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            groupId: imageGroupHandle === null || imageGroupHandle === void 0 ? void 0 : imageGroupHandle.groupId,
            previewInstRef,
            imageRef,
            imgProps: imgPropsRef,
            handleClick: () => {
                if (imageGroupHandle) {
                    imageGroupHandle.setPreviewSrc(props.src);
                    imageGroupHandle.setThumbnailEl(imageRef.value);
                    imageGroupHandle.toggleShow();
                    return;
                }
                const { value: previewInst } = previewInstRef;
                if (!previewInst)
                    return;
                previewInst.setPreviewSrc(props.src);
                previewInst.setThumbnailEl(imageRef.value);
                previewInst.toggleShow();
            }
        };
    },
    render() {
        const { mergedClsPrefix, imgProps = {} } = this;
        const imgWrapperNode = vue_1.h('div', vue_1.mergeProps(this.$attrs, {
            role: 'none',
            class: `${mergedClsPrefix}-image`
        }), vue_1.h("img", Object.assign({}, imgProps, { class: this.groupId, ref: "imageRef", width: this.width ? this.width : imgProps.width, height: this.height ? this.height : imgProps.height, src: this.src ? this.src : imgProps.src, alt: this.alt ? this.alt : imgProps.alt, "aria-label": this.alt ? this.alt : imgProps.alt, onClick: this.handleClick })));
        return this.groupId ? (imgWrapperNode) : (vue_1.h(ImagePreview_1.default, { clsPrefix: mergedClsPrefix, ref: "previewInstRef", showToolbar: this.showToolbar }, {
            default: () => imgWrapperNode
        }));
    }
});
