// naive-ui/lib/image/src/ImageGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.imageGroupInjectionKey = void 0;
const vue_1 = require("vue");
const seemly_1 = require("seemly");
const ImagePreview_1 = require("./ImagePreview");
const _mixins_1 = require("../../_mixins");
exports.imageGroupInjectionKey = Symbol('image-group');
const imageGroupProps = {
    showToolbar: { type: Boolean, default: true }
};
exports.default = vue_1.defineComponent({
    name: 'ImageGroup',
    props: imageGroupProps,
    setup(props) {
        let currentSrc;
        const { mergedClsPrefixRef } = _mixins_1.useConfig(props);
        const groupId = seemly_1.createId();
        const vm = vue_1.getCurrentInstance();
        const setPreviewSrc = (src) => {
            var _a;
            currentSrc = src;
            (_a = previewInstRef.value) === null || _a === void 0 ? void 0 : _a.setPreviewSrc(src);
        };
        function go(step) {
            if (!(vm === null || vm === void 0 ? void 0 : vm.proxy))
                return;
            const container = vm.proxy.$el.parentElement;
            // use dom api since we can't get the correct order before all children are rendered
            const imgs = container.getElementsByClassName(groupId);
            if (!imgs.length)
                return;
            const index = Array.from(imgs).findIndex((img) => img.src === currentSrc);
            if (~index) {
                setPreviewSrc(imgs[(index + step + imgs.length) % imgs.length].src);
            }
            else {
                setPreviewSrc(imgs[0].src);
            }
        }
        vue_1.provide(exports.imageGroupInjectionKey, {
            mergedClsPrefixRef,
            setPreviewSrc,
            setThumbnailEl: (el) => {
                var _a;
                (_a = previewInstRef.value) === null || _a === void 0 ? void 0 : _a.setThumbnailEl(el);
            },
            toggleShow: () => {
                var _a;
                (_a = previewInstRef.value) === null || _a === void 0 ? void 0 : _a.toggleShow();
            },
            groupId
        });
        const previewInstRef = vue_1.ref(null);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            previewInstRef,
            next: () => go(1),
            prev: () => go(-1)
        };
    },
    render() {
        return (vue_1.h(ImagePreview_1.default, { clsPrefix: this.mergedClsPrefix, ref: "previewInstRef", onPrev: this.prev, onNext: this.next, showToolbar: this.showToolbar }, {
            default: () => vue_1.renderSlot(this.$slots, 'default')
        }));
    }
});
