// naive-ui/lib/typography/src/a.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _mixins_1 = require("../../_mixins");
const styles_1 = require("../styles");
const a_cssr_1 = require("./styles/a.cssr");
const aProps = Object.assign({}, _mixins_1.useTheme.props);
exports.default = vue_1.defineComponent({
    name: 'A',
    props: aProps,
    setup(props) {
        const { mergedClsPrefixRef } = _mixins_1.useConfig(props);
        const themeRef = _mixins_1.useTheme('Typography', 'A', a_cssr_1.default, styles_1.typographyLight, props, mergedClsPrefixRef);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            cssVars: vue_1.computed(() => {
                const { common: { cubicBezierEaseInOut }, self: { aTextColor } } = themeRef.value;
                return {
                    '--text-color': aTextColor,
                    '--bezier': cubicBezierEaseInOut
                };
            })
        };
    },
    render() {
        return (vue_1.h("a", { class: `${this.mergedClsPrefix}-a`, style: this.cssVars }, this.$slots));
    }
});
