// naive-ui/lib/layout/src/LayoutFooter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _mixins_1 = require("../../_mixins");
const styles_1 = require("../styles");
const interface_1 = require("./interface");
const layout_footer_cssr_1 = require("./styles/layout-footer.cssr");
const layoutFooterProps = Object.assign(Object.assign({}, _mixins_1.useTheme.props), { inverted: Boolean, position: interface_1.positionProp, bordered: Boolean });
exports.default = vue_1.defineComponent({
    name: 'LayoutFooter',
    props: layoutFooterProps,
    setup(props) {
        const { mergedClsPrefixRef } = _mixins_1.useConfig(props);
        const themeRef = _mixins_1.useTheme('Layout', 'LayoutFooter', layout_footer_cssr_1.default, styles_1.layoutLight, props, mergedClsPrefixRef);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            cssVars: vue_1.computed(() => {
                const { common: { cubicBezierEaseInOut }, self } = themeRef.value;
                const vars = {
                    '--bezier': cubicBezierEaseInOut
                };
                if (props.inverted) {
                    vars['--color'] = self.footerColorInverted;
                    vars['--text-color'] = self.textColorInverted;
                    vars['--border-color'] = self.footerBorderColorInverted;
                }
                else {
                    vars['--color'] = self.footerColor;
                    vars['--text-color'] = self.textColor;
                    vars['--border-color'] = self.footerBorderColor;
                }
                return vars;
            })
        };
    },
    render() {
        const { mergedClsPrefix } = this;
        return (vue_1.h("div", { class: [
                `${mergedClsPrefix}-layout-footer`,
                this.position &&
                    `${mergedClsPrefix}-layout-footer--${this.position}-positioned`,
                this.bordered && `${mergedClsPrefix}-layout-footer--bordered`
            ], style: this.cssVars }, this.$slots));
    }
});
