// naive-ui/lib/collapse/src/CollapseItemContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const vooks_1 = require("vooks");
const _internal_1 = require("../../_internal");
exports.default = vue_1.defineComponent({
    name: 'CollapseItemContent',
    props: {
        displayDirective: {
            type: String,
            required: true
        },
        show: Boolean,
        clsPrefix: {
            type: String,
            required: true
        }
    },
    setup(props) {
        const onceTrueRef = vooks_1.useFalseUntilTruthy(vue_1.toRef(props, 'show'));
        return {
            onceTrue: onceTrueRef
        };
    },
    render() {
        return (vue_1.h(_internal_1.NFadeInExpandTransition, null, {
            default: () => {
                const { show, displayDirective, onceTrue, clsPrefix } = this;
                const useVShow = displayDirective === 'show' && onceTrue;
                const contentNode = (vue_1.h("div", { class: `${clsPrefix}-collapse-item__content-wrapper` },
                    vue_1.h("div", { class: `${clsPrefix}-collapse-item__content-inner` }, this.$slots)));
                return useVShow
                    ? vue_1.withDirectives(contentNode, [[vue_1.vShow, show]])
                    : show
                        ? contentNode
                        : null;
            }
        }));
    }
});
