// naive-ui/lib/tabs/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --bezier
// --close-color
// --close-color-hover
// --close-color-pressed
// --bar-color
// --tab-font-size
// --tab-text-color
// --tab-text-color-active
// --tab-text-color-disabled
// --tab-text-color-hover
// --pane-text-color
// --tab-border-color
// --tab-border-radius
// --tab-color
// --tab-font-weight
// --tab-font-weight-active
// --tab-gap
// --tab-padding
// --pane-padding


exports.default = cssr_1.cB('tabs', `
 width: 100%;
 transition:
 background-color .3s var(--bezier),
 border-color .3s var(--bezier);
`, [cssr_1.cM('flex', [cssr_1.cB('tabs-nav', {
  width: '100%'
}, [cssr_1.cB('tabs-wrapper', {
  width: '100%'
}, [cssr_1.cB('tabs-tab', {
  marginRight: 0
})])])]), cssr_1.cB('tabs-nav', `
 box-sizing: border-box;
 line-height: 1.5;
 display: flex;
 transition: border-color .3s var(--bezier);
 `, [cssr_1.cE('prefix, suffix', `
 display: flex;
 align-items: center;
 `), cssr_1.cE('prefix', 'padding-right: 16px;'), cssr_1.cE('suffix', 'padding-left: 16px;')]), cssr_1.cB('tabs-nav-scroll-wrapper', `
 flex: 1;
 position: relative;
 overflow: hidden;
 `, [cssr_1.cM('shadow-before', [cssr_1.c('&::before', `
 box-shadow: inset 10px 0 8px -8px rgba(0, 0, 0, .12);
 `)]), cssr_1.cM('shadow-after', [cssr_1.c('&::after', `
 box-shadow: inset -10px 0 8px -8px rgba(0, 0, 0, .12);
 `)]), cssr_1.c('&::before', `
 transition: box-shadow .3s var(--bezier);
 pointer-events: none;
 content: "";
 position: absolute;
 left: 0;
 top: 0;
 bottom: 0;
 width: 20px;
 `), cssr_1.c('&::after', `
 transition: box-shadow .3s var(--bezier);
 pointer-events: none;
 content: "";
 position: absolute;
 right: 0;
 top: 0;
 bottom: 0;
 width: 20px;
 `)]), cssr_1.cB('tabs-nav-scroll-content', `
 display: flex;
 position: relative;
 `), cssr_1.cB('tabs-wrapper', `
 display: inline-flex;
 flex-wrap: nowrap;
 position: relative;
 `), cssr_1.cB('tabs-tab-wrapper', `
 display: flex;
 flex-wrap: nowrap;
 flex-shrink: 0;
 flex-grow: 0;
 `), cssr_1.cB('tabs-tab', `
 cursor: pointer;
 white-space: nowrap;
 flex-wrap: nowrap;
 display: inline-flex;
 align-items: center;
 color: var(--tab-text-color);
 font-size: var(--tab-font-size);
 background-clip: padding-box;
 padding: var(--tab-padding);
 transition:
 color .3s var(--bezier),
 background-color .3s var(--bezier),
 border-color .3s var(--bezier);
 `, [cssr_1.cM('disabled', {
  cursor: 'not-allowed'
}), cssr_1.cE('close', `
 margin-left: 8px;
 font-size: 14px;
 transition: color .3s var(--bezier);
 `), cssr_1.cE('label', `
 display: flex;
 align-items: center;
 `)]), cssr_1.cB('tabs-bar', `
 position: absolute;
 bottom: 0;
 height: 2px;
 border-radius: 1px;
 background-color: var(--bar-color);
 transition:
 left .2s var(--bezier),
 max-width .2s var(--bezier),
 background-color .3s var(--bezier);
 `, [cssr_1.cM('transition-disabled', `
 transition: none;
 `), cssr_1.cM('disabled', `
 background-color: var(--tab-text-color-disabled)
 `)]), cssr_1.cB('tab-pane', `
 color: var(--pane-text-color);
 width: 100%;
 padding: var(--pane-padding);
 transition:
 color .3s var(--bezier),
 background-color .3s var(--bezier);
 `), cssr_1.cB('tabs-tab-pad', `
 width: var(--tab-gap);
 flex-grow: 0;
 flex-shrink: 0;
 `), cssr_1.cM('line-type, bar-type', [cssr_1.cB('tabs-tab', `
 font-weight: var(--tab-font-weight-active);
 box-sizing: border-box;
 vertical-align: bottom;
 `, [cssr_1.c('&:hover', {
  color: 'var(--tab-text-color-hover)'
}), cssr_1.cM('active', {
  color: 'var(--tab-text-color-active)'
}), cssr_1.cM('disabled', {
  color: 'var(--tab-text-color-disabled)'
})])]), cssr_1.cM('line-type', [cssr_1.cB('tabs-nav', [cssr_1.cE('prefix, suffix', `
 transition: border-color .3s var(--bezier);
 border-bottom: 1px solid var(--tab-border-color);
 `)]), cssr_1.cB('tabs-nav-scroll-content', `
 transition: border-color .3s var(--bezier);
 border-bottom: 1px solid var(--tab-border-color);
 `), cssr_1.cB('tabs-bar', `
 border-radius: 0;
 bottom: -1px;
 `)]), cssr_1.cM('card-type', [cssr_1.cB('tabs-nav', [cssr_1.cE('prefix, suffix', `
 transition: border-color .3s var(--bezier);
 border-bottom: 1px solid var(--tab-border-color);
 `)]), cssr_1.cB('tabs-pad', `
 flex-grow: 1;
 transition: border-color .3s var(--bezier);
 border-bottom: 1px solid var(--tab-border-color);
 `), cssr_1.cB('tabs-tab-pad', `
 transition: border-color .3s var(--bezier);
 border-bottom: 1px solid var(--tab-border-color);
 `), cssr_1.cB('tabs-tab', `
 font-weight: var(--tab-font-weight);
 border: 1px solid var(--tab-border-color);
 border-top-left-radius: var(--tab-border-radius);
 border-top-right-radius: var(--tab-border-radius);
 background-color: var(--tab-color);
 box-sizing: border-box;
 position: relative;
 vertical-align: bottom;
 display: flex;
 justify-content: space-between;
 font-size: var(--tab-font-size);
 color: var(--tab-text-color);
 `, [cssr_1.cM('addable', `
 padding-left: 8px;
 padding-right: 8px;
 font-size: 16px;
 `, [cssr_1.cNotM('disabled', [cssr_1.c('&:hover', `
 color: var(--tab-text-color-active);
 `)])]), cssr_1.cM('closable', 'padding-right: 6px;'), cssr_1.cM('active', `
 border-bottom: 1px solid #0000;
 background-color: #0000;
 font-weight: var(--tab-font-weight-active);
 color: var(--tab-text-color-active);
 `), cssr_1.cM('disabled', 'color: var(--tab-text-color-disabled);')]), cssr_1.cB('tabs-scroll-padding', 'border-bottom: 1px solid var(--tab-border-color);')])]);
