// naive-ui/lib/cascader/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_in_scale_up_cssr_1 = require("../../../_styles/transitions/fade-in-scale-up.cssr"); // vars:
// --menu-border-radius
// --menu-box-shadow
// --menu-height
// --menu-color
// --menu-divider-color
// --option-height
// --option-font-size
// --option-text-color
// --option-text-color-disabled
// --option-text-color-active
// --option-color-hover
// --option-check-mark-color
// --option-arrow-color
// --menu-mask-color
// --loading-color


exports.default = cssr_1.c([cssr_1.cB('cascader-menu', `
 outline: none;
 position: relative;
 margin: 4px 0;
 display: flex;
 flex-flow: column nowrap;
 border-radius: var(--menu-border-radius);
 overflow: hidden;
 box-shadow: var(--menu-box-shadow);
 color: var(--option-text-color);
 background-color: var(--menu-color);
 `, [fade_in_scale_up_cssr_1.default({
  transformOrigin: 'inherit',
  duration: '0.2s'
}), cssr_1.cB('scrollbar', {
  // if width not set, cascader select menu's inner scroll area's width is
  // not correct, which won't change after select menu width is set
  width: '100%'
}), cssr_1.cB('base-menu-mask', {
  backgroundColor: 'var(--menu-mask-color)'
}), cssr_1.cB('base-loading', {
  color: 'var(--loading-color)'
}), cssr_1.cB('cascader-submenu-wrapper', `
 position: relative;
 display: flex;
 flex-wrap: nowrap;
 `), cssr_1.cB('cascader-submenu', `
 height: var(--menu-height);
 position: relative;
 min-width: 180px;
 `, [cssr_1.cB('scrollbar-content', {
  position: 'relative'
}), cssr_1.c('&:first-child', `
 border-top-left-radius: var(--menu-border-radius);
 border-bottom-left-radius: var(--menu-border-radius);
 `), cssr_1.c('&:last-child', `
 border-top-right-radius: var(--menu-border-radius);
 border-bottom-right-radius: var(--menu-border-radius);
 `), cssr_1.c('&:not(:first-child)', `
 border-left: 1px solid var(--menu-divider-color);
 `)]), cssr_1.cB('cascader-menu-action', `
 box-sizing: border-box;
 padding: 8px;
 border-top: 1px solid var(--menu-divider-color);
 `), cssr_1.cB('cascader-option', `
 height: var(--option-height);
 line-height: var(--option-height);
 font-size: var(--option-font-size);
 padding: 0 0 0 18px;
 box-sizing: border-box;
 min-width: 182px;
 background-color: #0000;
 display: flex;
 align-items: center;
 white-space: nowrap;
 position: relative;
 cursor: pointer;
 transition:
 background-color .2s var(--bezier),
 color 0.2s var(--bezier);
 `, [cssr_1.cM('show-prefix', {
  paddingLeft: 0
}), cssr_1.cE('label', `
 flex: 1 0 0;
 overflow: hidden;
 text-overflow: ellipsis;
 `), cssr_1.cE('prefix', {
  width: '32px',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center'
}), cssr_1.cE('suffix', {
  width: '32px',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center'
}), cssr_1.cB('cascader-option-icon-placeholder', {
  lineHeight: 0,
  position: 'relative',
  width: '16px',
  height: '16px',
  fontSize: '16px'
}, [cssr_1.cB('cascader-option-icon', [cssr_1.cM('checkmark', {
  color: 'var(--option-check-mark-color)'
}, [fade_in_scale_up_cssr_1.default()]), cssr_1.cM('arrow', {
  color: 'var(--option-arrow-color)'
})])]), cssr_1.cM('selected', {
  color: 'var(--option-text-color-active)'
}), cssr_1.cM('active', {
  color: 'var(--option-text-color-active)',
  backgroundColor: 'var(--option-color-hover)'
}), cssr_1.cM('pending', {
  backgroundColor: 'var(--option-color-hover)'
}), cssr_1.c('&:hover', {
  backgroundColor: 'var(--option-color-hover)'
}), cssr_1.cM('disabled', `
 color: var(--option-text-color-disabled);
 background-color: #0000;
 cursor: not-allowed;
 `, [cssr_1.cB('cascader-option-icon', [cssr_1.cM('arrow', {
  color: 'var(--option-text-color-disabled)'
})])])])]), cssr_1.cB('cascader', `
 z-index: auto;
 position: relative;
 width: 100%;
 `)]);
