// naive-ui/lib/drawer/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const slide_in_from_right_1 = require("../../../_styles/transitions/slide-in-from-right");

const slide_in_from_left_1 = require("../../../_styles/transitions/slide-in-from-left");

const slide_in_from_top_1 = require("../../../_styles/transitions/slide-in-from-top");

const slide_in_from_bottom_1 = require("../../../_styles/transitions/slide-in-from-bottom");

const fade_in_cssr_1 = require("../../../_styles/transitions/fade-in.cssr"); // vars:
// --line-height
// --color
// --text-color
// --box-shadow
// --bezier
// --bezier-out
// --bezier-in
// --body-padding
// --header-padding
// --footer-padding
// --title-font-size
// --title-text-color
// --title-font-weight
// --header-border-bottom
// --footer-border-top
// --close-color
// --close-color-hover
// --close-color-pressed
// --close-size


exports.default = cssr_1.c([cssr_1.cB('drawer', `
 line-height: var(--line-height);
 position: absolute;
 pointer-events: all;
 box-shadow: var(--box-shadow);
 transition:
 background-color .3s var(--bezier),
 color .3s var(--bezier);
 background-color: var(--color);
 color: var(--text-color);
 box-sizing: border-box;
 `, [slide_in_from_right_1.default(), slide_in_from_left_1.default(), slide_in_from_top_1.default(), slide_in_from_bottom_1.default(), cssr_1.cM('native-scrollbar', [cssr_1.cB('drawer-content-wrapper', `
 overflow: auto;
 height: 100%;
 `)]), cssr_1.cB('drawer-content-wrapper', `
 box-sizing: border-box;
 `), cssr_1.cB('drawer-content', `
 height: 100%;
 display: flex;
 flex-direction: column;
 `, [cssr_1.cM('native-scrollbar', [cssr_1.cB('drawer-body-content-wrapper', `
 height: 100%;
 overflow: auto;
 `)]), cssr_1.cB('drawer-body', `
 flex: 1 0 0;
 `), cssr_1.cB('drawer-body-content-wrapper', `
 box-sizing: border-box;
 padding: var(--body-padding);
 `), cssr_1.cB('drawer-header', `
 font-weight: var(--title-font-weight);
 line-height: 1;
 font-size: var(--title-font-size);
 color: var(--title-text-color);
 padding: var(--header-padding);
 border-bottom: 1px solid var(--divider-color);
 border-bottom: var(--header-border-bottom);
 display: flex;
 justify-content: space-between;
 align-items: center;
 `, [cssr_1.cE('close', `
 transition: color .3s var(--bezier);
 font-size: var(--close-size);
 `)]), cssr_1.cB('drawer-footer', `
 display: flex;
 justify-content: flex-end;
 border-top: var(--footer-border-top);
 padding: var(--footer-padding);
 `)]), cssr_1.cM('right-placement', `
 top: 0;
 bottom: 0;
 right: 0;
 `), cssr_1.cM('left-placement', `
 top: 0;
 bottom: 0;
 left: 0;
 `), cssr_1.cM('top-placement', `
 top: 0;
 left: 0;
 right: 0;
 `), cssr_1.cM('bottom-placement', `
 left: 0;
 bottom: 0;
 right: 0;
 `)]), cssr_1.c('body', [cssr_1.c('>', [cssr_1.cB('drawer-container', {
  position: 'fixed'
})])]), cssr_1.cB('drawer-container', `
 position: relative;
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 pointer-events: none;
 `, [cssr_1.c('> *', {
  pointerEvents: 'all'
})]), cssr_1.cB('drawer-mask', `
 background-color: rgba(0, 0, 0, .3);
 position: absolute;
 left: 0;
 right: 0;
 top: 0;
 bottom: 0;
 `, [fade_in_cssr_1.default({
  enterDuration: '0.2s',
  leaveDuration: '0.2s',
  enterCubicBezier: 'var(--bezier-in)',
  leaveCubicBezier: 'var(--bezier-out)'
})])]);
